{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries / Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biomart import BiomartServer\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import gtfparse\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo sapiens\n",
      "DictElement({'RsUid': '', 'GbUid': '47542708', 'AssemblyAccession': 'GCA_963426525.1', 'LastMajorReleaseAccession': 'GCA_963426525.1', 'LatestAccession': '', 'ChainId': '963426525', 'AssemblyName': 'human feces metagenome', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '9606', 'Organism': 'Homo sapiens (human)', 'SpeciesTaxid': '9606', 'SpeciesName': 'Homo sapiens', 'AssemblyType': 'haploid', 'AssemblyStatus': 'Contig', 'AssemblyStatusSort': '6', 'WGS': 'CAUKAQ01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJEB65255', 'BioprojectId': '1014690'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMEA114333763', 'BioSampleId': '37336560', 'Biosource': {'InfraspeciesList': [], 'Sex': '', 'Isolate': ''}, 'Coverage': '31', 'PartialGenomeRepresentation': 'false', 'Primary': '47542698', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2023/09/16 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2023/09/13 00:00', 'AsmUpdateDate': '2023/09/16 00:00', 'SubmissionDate': '2023/09/13 00:00', 'LastUpdateDate': '2023/09/16 00:00', 'SubmitterOrganization': 'University of Oslo', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': ['genome length too small'], 'PropertyList': ['excluded-from-refseq', 'full-genome-representation', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_963426525.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '58898', 'ScaffoldN50': '58898', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '9C6X01019999411019634265259800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">1475</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">411</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">58898</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">1475</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">1475</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">411</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">58898</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">78719706</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">78719706</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>1</assembly-level> <assembly-status>Contig</assembly-status> <representative-status>na</representative-status> <submitter-organization>University of Oslo</submitter-organization>    '}, attributes={'uid': '18601811'})\n",
      "Homo sapiens: DictElement({'RsUid': '', 'GbUid': '47542708', 'AssemblyAccession': 'GCA_963426525.1', 'LastMajorReleaseAccession': 'GCA_963426525.1', 'LatestAccession': '', 'ChainId': '963426525', 'AssemblyName': 'human feces metagenome', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '9606', 'Organism': 'Homo sapiens (human)', 'SpeciesTaxid': '9606', 'SpeciesName': 'Homo sapiens', 'AssemblyType': 'haploid', 'AssemblyStatus': 'Contig', 'AssemblyStatusSort': '6', 'WGS': 'CAUKAQ01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJEB65255', 'BioprojectId': '1014690'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMEA114333763', 'BioSampleId': '37336560', 'Biosource': {'InfraspeciesList': [], 'Sex': '', 'Isolate': ''}, 'Coverage': '31', 'PartialGenomeRepresentation': 'false', 'Primary': '47542698', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2023/09/16 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2023/09/13 00:00', 'AsmUpdateDate': '2023/09/16 00:00', 'SubmissionDate': '2023/09/13 00:00', 'LastUpdateDate': '2023/09/16 00:00', 'SubmitterOrganization': 'University of Oslo', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': ['genome length too small'], 'PropertyList': ['excluded-from-refseq', 'full-genome-representation', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_963426525.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '58898', 'ScaffoldN50': '58898', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '9C6X01019999411019634265259800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">1475</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">411</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">58898</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">1475</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">1475</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">411</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">58898</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">78719706</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">78719706</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/963/426/525/GCA_963426525.1_human_feces_metagenome/GCA_963426525.1_human_feces_metagenome_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>1</assembly-level> <assembly-status>Contig</assembly-status> <representative-status>na</representative-status> <submitter-organization>University of Oslo</submitter-organization>    '}, attributes={'uid': '18601811'})\n",
      "Mus musculus\n",
      "DictElement({'RsUid': '', 'GbUid': '47846178', 'AssemblyAccession': 'GCA_031761455.1', 'LastMajorReleaseAccession': 'GCA_031761455.1', 'LatestAccession': '', 'ChainId': '31761455', 'AssemblyName': 'BALBC_Nude', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '10090', 'Organism': 'Mus musculus (house mouse)', 'SpeciesTaxid': '10090', 'SpeciesName': 'Mus musculus', 'AssemblyType': 'haploid', 'AssemblyStatus': 'Chromosome', 'AssemblyStatusSort': '3', 'WGS': 'JARJHH01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJNA944543', 'BioprojectId': '944543'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMN33748820', 'BioSampleId': '33748820', 'Biosource': {'InfraspeciesList': [{'Sub_type': 'strain', 'Sub_value': 'BALB/c Nude'}], 'Sex': 'female', 'Isolate': ''}, 'Coverage': '45', 'PartialGenomeRepresentation': 'false', 'Primary': '47845718', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2023/09/22 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2023/09/22 00:00', 'AsmUpdateDate': '2023/09/22 00:00', 'SubmissionDate': '2023/09/22 00:00', 'LastUpdateDate': '2023/09/22 00:00', 'SubmitterOrganization': 'JSR Life Sciences', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': [], 'PropertyList': ['full-genome-representation', 'has-chromosome', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_031761455.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '30118904', 'ScaffoldN50': '101020256', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '5C3X02029698810950317614559800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">20</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">1224</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">28</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">30118904</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">20</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">1089</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">12</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">19</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">1058</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">11</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">101020256</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">2754161322</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">2745714133</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>3</assembly-level> <assembly-status>Chromosome</assembly-status> <representative-status>na</representative-status> <submitter-organization>JSR Life Sciences</submitter-organization>    '}, attributes={'uid': '18738731'})\n",
      "Mus musculus: DictElement({'RsUid': '', 'GbUid': '47846178', 'AssemblyAccession': 'GCA_031761455.1', 'LastMajorReleaseAccession': 'GCA_031761455.1', 'LatestAccession': '', 'ChainId': '31761455', 'AssemblyName': 'BALBC_Nude', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '10090', 'Organism': 'Mus musculus (house mouse)', 'SpeciesTaxid': '10090', 'SpeciesName': 'Mus musculus', 'AssemblyType': 'haploid', 'AssemblyStatus': 'Chromosome', 'AssemblyStatusSort': '3', 'WGS': 'JARJHH01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJNA944543', 'BioprojectId': '944543'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMN33748820', 'BioSampleId': '33748820', 'Biosource': {'InfraspeciesList': [{'Sub_type': 'strain', 'Sub_value': 'BALB/c Nude'}], 'Sex': 'female', 'Isolate': ''}, 'Coverage': '45', 'PartialGenomeRepresentation': 'false', 'Primary': '47845718', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2023/09/22 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2023/09/22 00:00', 'AsmUpdateDate': '2023/09/22 00:00', 'SubmissionDate': '2023/09/22 00:00', 'LastUpdateDate': '2023/09/22 00:00', 'SubmitterOrganization': 'JSR Life Sciences', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': [], 'PropertyList': ['full-genome-representation', 'has-chromosome', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_031761455.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '30118904', 'ScaffoldN50': '101020256', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '5C3X02029698810950317614559800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">20</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">1224</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">28</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">30118904</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">20</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">1089</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">12</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">19</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">1058</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">11</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">101020256</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">2754161322</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">2745714133</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/031/761/455/GCA_031761455.1_BALBC_Nude/GCA_031761455.1_BALBC_Nude_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>3</assembly-level> <assembly-status>Chromosome</assembly-status> <representative-status>na</representative-status> <submitter-organization>JSR Life Sciences</submitter-organization>    '}, attributes={'uid': '18738731'})\n",
      "Pan paniscus\n",
      "DictElement({'RsUid': '', 'GbUid': '43265948', 'AssemblyAccession': 'GCA_030221875.1', 'LastMajorReleaseAccession': 'GCA_030221875.1', 'LatestAccession': '', 'ChainId': '30221875', 'AssemblyName': 'Mhudiblu_PPA_hifiasm-v0.15.2.pri', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '9597', 'Organism': 'Pan paniscus (pygmy chimpanzee)', 'SpeciesTaxid': '9597', 'SpeciesName': 'Pan paniscus', 'AssemblyType': 'haploid (principal pseudohaplotype of diploid)', 'AssemblyStatus': 'Contig', 'AssemblyStatusSort': '6', 'WGS': 'JARRAW01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJNA941363', 'BioprojectId': '941363'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMN30617506', 'BioSampleId': '30617506', 'Biosource': {'InfraspeciesList': [], 'Sex': 'female', 'Isolate': 'Mhudiblu'}, 'Coverage': '39', 'PartialGenomeRepresentation': 'false', 'Primary': '43265938', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2023/06/07 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2023/06/07 00:00', 'AsmUpdateDate': '2023/06/07 00:00', 'SubmissionDate': '2023/06/07 00:00', 'LastUpdateDate': '2023/06/07 00:00', 'SubmitterOrganization': 'University of Washington', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': [], 'PropertyList': ['full-genome-representation', 'has_linked_assembly', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_030221875.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '50451699', 'ScaffoldN50': '50451699', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '5C6X9F9F9495483000302218759800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">938</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">21</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">50451699</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">938</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">938</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">21</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">50451699</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">3189067193</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">3189067193</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>1</assembly-level> <assembly-status>Contig</assembly-status> <representative-status>na</representative-status> <submitter-organization>University of Washington</submitter-organization>    '}, attributes={'uid': '16840231'})\n",
      "Pan paniscus: DictElement({'RsUid': '', 'GbUid': '43265948', 'AssemblyAccession': 'GCA_030221875.1', 'LastMajorReleaseAccession': 'GCA_030221875.1', 'LatestAccession': '', 'ChainId': '30221875', 'AssemblyName': 'Mhudiblu_PPA_hifiasm-v0.15.2.pri', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '9597', 'Organism': 'Pan paniscus (pygmy chimpanzee)', 'SpeciesTaxid': '9597', 'SpeciesName': 'Pan paniscus', 'AssemblyType': 'haploid (principal pseudohaplotype of diploid)', 'AssemblyStatus': 'Contig', 'AssemblyStatusSort': '6', 'WGS': 'JARRAW01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJNA941363', 'BioprojectId': '941363'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMN30617506', 'BioSampleId': '30617506', 'Biosource': {'InfraspeciesList': [], 'Sex': 'female', 'Isolate': 'Mhudiblu'}, 'Coverage': '39', 'PartialGenomeRepresentation': 'false', 'Primary': '43265938', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2023/06/07 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2023/06/07 00:00', 'AsmUpdateDate': '2023/06/07 00:00', 'SubmissionDate': '2023/06/07 00:00', 'LastUpdateDate': '2023/06/07 00:00', 'SubmitterOrganization': 'University of Washington', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': [], 'PropertyList': ['full-genome-representation', 'has_linked_assembly', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_030221875.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '50451699', 'ScaffoldN50': '50451699', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '5C6X9F9F9495483000302218759800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">938</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">21</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">50451699</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">938</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">938</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">21</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">50451699</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">3189067193</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">3189067193</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/221/875/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri/GCA_030221875.1_Mhudiblu_PPA_hifiasm-v0.15.2.pri_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>1</assembly-level> <assembly-status>Contig</assembly-status> <representative-status>na</representative-status> <submitter-organization>University of Washington</submitter-organization>    '}, attributes={'uid': '16840231'})\n",
      "not animal\n",
      "Warning: Organism from NCBI ('Suid alphaherpesvirus 1 (viruses)') does not closely match the input species name ('not animal')!\n",
      "not animal: None\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from difflib import SequenceMatcher\n",
    "Entrez.email = \"matthew.schmitz@alleninstitute.org\"\n",
    "\n",
    "def get_ncbi_assembly_stats(species_name):\n",
    "    handle = Entrez.esearch(db=\"assembly\", term=species_name)\n",
    "    record = Entrez.read(handle)\n",
    "    assembly_id = record['IdList'][0]\n",
    "    summary_handle = Entrez.esummary(db=\"assembly\", id=assembly_id)\n",
    "    summary = Entrez.read(summary_handle)\n",
    "    \n",
    "    organism_from_ncbi = summary['DocumentSummarySet']['DocumentSummary'][0]['Organism']\n",
    "    match_ratio = SequenceMatcher(None, organism_from_ncbi, species_name).ratio()\n",
    "\n",
    "    # Check if the match ratio is above a certain threshold\n",
    "    # You can adjust this threshold as needed\n",
    "    if match_ratio < 0.4:\n",
    "        print(f\"Warning: Organism from NCBI ('{organism_from_ncbi}') does not closely match the input species name ('{species_name}')!\")\n",
    "        return None\n",
    "\n",
    "    print(summary['DocumentSummarySet']['DocumentSummary'][0])\n",
    "    return summary['DocumentSummarySet']['DocumentSummary'][0]\n",
    "\n",
    "species_list = ['Homo sapiens', 'Mus musculus', 'Pan paniscus', 'not animal']\n",
    "for species in species_list:\n",
    "    print(species)\n",
    "    print(f\"{species}: {get_ncbi_assembly_stats(species)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "homo sapiens\n",
      "{'AssemblyName': 'GRCh38.p14', 'AssemblyAccession': 'GCA_000001405.29', 'SpeciesName': 'Homo sapiens', 'AssemblyType': 'diploid', 'TotalLength': 3099750718, 'ChromosomeCount': 25, 'ContigCount': 0, 'ScaffoldCount': 169, 'AssemblyStatus': 'Chromosome', 'GenebuildLastUpdate': '2023-03'}\n",
      "mus musculus\n",
      "{'AssemblyName': 'GRCm39', 'AssemblyAccession': 'GCA_000001635.9', 'SpeciesName': 'Mus musculus', 'AssemblyType': 'diploid', 'TotalLength': 2728222451, 'ChromosomeCount': 22, 'ContigCount': 0, 'ScaffoldCount': 39, 'AssemblyStatus': 'Chromosome', 'GenebuildLastUpdate': '2023-04'}\n",
      "pan paniscus\n",
      "{'AssemblyName': 'panpan1.1', 'AssemblyAccession': 'GCA_000258655.2', 'SpeciesName': 'Pan paniscus', 'AssemblyType': 'diploid', 'TotalLength': 3286643896, 'ChromosomeCount': 25, 'ContigCount': 0, 'ScaffoldCount': 10249, 'AssemblyStatus': 'Chromosome', 'GenebuildLastUpdate': '2020-03'}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def get_ensembl_assembly_stats(species_name):\n",
    "    url = f\"https://rest.ensembl.org/info/assembly/{species_name}?content-type=application/json\"\n",
    "    response = requests.get(url)\n",
    "    data = response.json()\n",
    "\n",
    "    relevant_info = {\n",
    "        'AssemblyName': data.get('assembly_name', ''),\n",
    "        'AssemblyAccession': data.get('assembly_accession', ''),\n",
    "        'SpeciesName': species_name.replace('_', ' ').capitalize(),\n",
    "        'AssemblyType': 'haploid' if len(data.get('karyotype', [])) == 1 else 'diploid',\n",
    "        'TotalLength': sum([region['length'] for region in data.get('top_level_region', [])]),\n",
    "        'ChromosomeCount': len(data.get('karyotype', [])),\n",
    "        'ContigCount': len([region for region in data.get('top_level_region', []) if region['coord_system'] == 'contig']),\n",
    "        'ScaffoldCount': len([region for region in data.get('top_level_region', []) if region['coord_system'] == 'scaffold']),\n",
    "        # Note: N50 stats would typically require a more detailed breakdown of scaffold/contig lengths which isn't immediately provided here\n",
    "        'AssemblyStatus': 'Chromosome' if any([region['coord_system'] == 'chromosome' for region in data.get('top_level_region', [])]) else 'Scaffold',\n",
    "        'GenebuildLastUpdate': data.get('genebuild_last_geneset_update', ''),\n",
    "    }\n",
    "\n",
    "    # This is just a subset of the fields you provided from NCBI. Some fields, like 'Coverage', 'ContigN50', and 'ScaffoldN50', are not immediately available from the Ensembl API (or at least, not from this endpoint).\n",
    "\n",
    "    return relevant_info\n",
    "\n",
    "species_list = ['homo_sapiens', 'mus_musculus', 'pan_paniscus']\n",
    "for species in species_list:\n",
    "    print(species.replace(\"_\", \" \"))\n",
    "    print(get_ensembl_assembly_stats(species))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'221 Goodbye.'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from ftplib import FTP\n",
    "\n",
    "# Define the species of interest\n",
    "species_list = [\"homo_sapiens\", \"mus_musculus\", \"danio_rerio\"]\n",
    "\n",
    "ensembl_base = \"ftp.ensembl.org\"\n",
    "ensembl_path = \"/pub/current_gtf/\"\n",
    "local_path='/home/matthew.schmitz/Matthew/genome/'\n",
    "\n",
    "\n",
    "ftp = FTP(ensembl_base)\n",
    "ftp.login()\n",
    "\n",
    "for species in species_list:\n",
    "    species_dir = ensembl_path + species + \"/\"\n",
    "    \n",
    "    # Get a list of files in the species directory\n",
    "    files = ftp.nlst(species_dir)\n",
    "    \n",
    "    # Identify the GTF file (assuming there's only one .gtf.gz per species directory)\n",
    "    gtf_file = next((f for f in files if f.endswith(\".gtf.gz\")), None)\n",
    "    \n",
    "    if gtf_file:\n",
    "        local_file_name = gtf_file.split('/')[-1]\n",
    "        with open(os.path.join(local_path,'ensembl',species,local_file_name), 'wb') as local_file:\n",
    "            ftp.retrbinary('RETR ' + gtf_file, local_file.write)\n",
    "\n",
    "ftp.quit()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "ename": "error_perm",
     "evalue": "550 Failed to open file.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31merror_perm\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 20\u001b[0m\n\u001b[1;32m     17\u001b[0m     local_file_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(local_path,\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mspecies\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.CORE.gtf.gz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     19\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(local_file_path, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m local_file:\n\u001b[0;32m---> 20\u001b[0m         ftp\u001b[38;5;241m.\u001b[39mretrbinary(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRETR \u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m ensembl_path \u001b[38;5;241m+\u001b[39m file_name, local_file\u001b[38;5;241m.\u001b[39mwrite)\n\u001b[1;32m     22\u001b[0m ftp\u001b[38;5;241m.\u001b[39mquit()\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mftplib\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m FTP\n",
      "File \u001b[0;32m/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/utils/miniconda3/envs/genomes/lib/python3.11/ftplib.py:436\u001b[0m, in \u001b[0;36mFTP.retrbinary\u001b[0;34m(self, cmd, callback, blocksize, rest)\u001b[0m\n\u001b[1;32m    422\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Retrieve data in binary mode.  A new port is created for you.\u001b[39;00m\n\u001b[1;32m    423\u001b[0m \n\u001b[1;32m    424\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    433\u001b[0m \u001b[38;5;124;03m  The response code.\u001b[39;00m\n\u001b[1;32m    434\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    435\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvoidcmd(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTYPE I\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 436\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtransfercmd(cmd, rest) \u001b[38;5;28;01mas\u001b[39;00m conn:\n\u001b[1;32m    437\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    438\u001b[0m         data \u001b[38;5;241m=\u001b[39m conn\u001b[38;5;241m.\u001b[39mrecv(blocksize)\n",
      "File \u001b[0;32m/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/utils/miniconda3/envs/genomes/lib/python3.11/ftplib.py:393\u001b[0m, in \u001b[0;36mFTP.transfercmd\u001b[0;34m(self, cmd, rest)\u001b[0m\n\u001b[1;32m    391\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mtransfercmd\u001b[39m(\u001b[38;5;28mself\u001b[39m, cmd, rest\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    392\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Like ntransfercmd() but returns only the socket.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 393\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mntransfercmd(cmd, rest)[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/utils/miniconda3/envs/genomes/lib/python3.11/ftplib.py:359\u001b[0m, in \u001b[0;36mFTP.ntransfercmd\u001b[0;34m(self, cmd, rest)\u001b[0m\n\u001b[1;32m    357\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m rest \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    358\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msendcmd(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mREST \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m rest)\n\u001b[0;32m--> 359\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msendcmd(cmd)\n\u001b[1;32m    360\u001b[0m \u001b[38;5;66;03m# Some servers apparently send a 200 reply to\u001b[39;00m\n\u001b[1;32m    361\u001b[0m \u001b[38;5;66;03m# a LIST or STOR command, before the 150 reply\u001b[39;00m\n\u001b[1;32m    362\u001b[0m \u001b[38;5;66;03m# (and way before the 226 reply). This seems to\u001b[39;00m\n\u001b[1;32m    363\u001b[0m \u001b[38;5;66;03m# be in violation of the protocol (which only allows\u001b[39;00m\n\u001b[1;32m    364\u001b[0m \u001b[38;5;66;03m# 1xx or error messages for LIST), so we just discard\u001b[39;00m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;66;03m# this response.\u001b[39;00m\n\u001b[1;32m    366\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m resp[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m2\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "File \u001b[0;32m/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/utils/miniconda3/envs/genomes/lib/python3.11/ftplib.py:281\u001b[0m, in \u001b[0;36mFTP.sendcmd\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    279\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m'''Send a command and return the response.'''\u001b[39;00m\n\u001b[1;32m    280\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mputcmd(cmd)\n\u001b[0;32m--> 281\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgetresp()\n",
      "File \u001b[0;32m/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/utils/miniconda3/envs/genomes/lib/python3.11/ftplib.py:254\u001b[0m, in \u001b[0;36mFTP.getresp\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    252\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m error_temp(resp)\n\u001b[1;32m    253\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m c \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m5\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m--> 254\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m error_perm(resp)\n\u001b[1;32m    255\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m error_proto(resp)\n",
      "\u001b[0;31merror_perm\u001b[0m: 550 Failed to open file."
     ]
    }
   ],
   "source": [
    "import requests\n",
    "from ftplib import FTP\n",
    "\n",
    "# Define the species of interest\n",
    "species_list = [\"homo_sapiens\", \"mus_musculus\", \"danio_rerio\"]\n",
    "local_path='/home/matthew.schmitz/Matthew/genome/'\n",
    "\n",
    "# Fetching from Ensembl\n",
    "ensembl_base = \"ftp.ensembl.org\"\n",
    "ensembl_path = \"/pub/current_gtf/\"\n",
    "\n",
    "ftp = FTP(ensembl_base)\n",
    "ftp.login()\n",
    "\n",
    "for species in species_list:\n",
    "    file_name = f\"{species}/{species}.CORE.gtf.gz\"\n",
    "    local_file_path = os.path.join(local_path,f\"{species}.CORE.gtf.gz\")\n",
    "    \n",
    "    with open(local_file_path, 'wb') as local_file:\n",
    "        ftp.retrbinary('RETR ' + ensembl_path + file_name, local_file.write)\n",
    "\n",
    "ftp.quit()\n",
    "\n",
    "from ftplib import FTP\n",
    "local_path='/home/matthew.schmitz/Matthew/genome/'\n",
    "\n",
    "# Define the species of interest\n",
    "species_list = [\"homo_sapiens\", \"mus_musculus\", \"danio_rerio\"]\n",
    "\n",
    "ensembl_base = \"ftp.ensembl.org\"\n",
    "ensembl_path = \"/pub/current_gtf/\"\n",
    "\n",
    "ftp = FTP(ensembl_base)\n",
    "ftp.login()\n",
    "\n",
    "for species in species_list:\n",
    "    species_dir = ensembl_path + species + \"/\"\n",
    "    \n",
    "    # Get a list of files in the species directory\n",
    "    files = ftp.nlst(species_dir)\n",
    "    \n",
    "    # Identify the GTF file (assuming there's only one .gtf.gz per species directory)\n",
    "    gtf_file = next((f for f in files if f.endswith(\".gtf.gz\")), None)\n",
    "    \n",
    "    if gtf_file:\n",
    "        local_file_name =  os.path.join(local_path,gtf_file.split('/')[-1])\n",
    "        with open(local_file_name, 'wb') as local_file:\n",
    "            ftp.retrbinary('RETR ' + gtf_file, local_file.write)\n",
    "\n",
    "ftp.quit()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def get_common_name(scientific_name):\n",
    "    url = f\"https://rest.ensembl.org/info/species?content-type=application/json\"\n",
    "    response = requests.get(url)\n",
    "    species_list = response.json()['species']\n",
    "\n",
    "    for species in species_list:\n",
    "        if species['name'] == scientific_name:\n",
    "            return species['common_name']\n",
    "\n",
    "    return None\n",
    "\n",
    "# Example usage\n",
    "species_names = [\"homo_sapiens\", \"mus_musculus\"]\n",
    "common_names = {}\n",
    "\n",
    "for name in species_names:\n",
    "    common_name = get_common_name(name)\n",
    "    if common_name:\n",
    "        common_names[name] = common_name\n",
    "    else:\n",
    "        print(f\"Couldn't find common name for {name}\")\n",
    "\n",
    "print(common_names)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map gene symbol to ensemble ID using gtf files\n",
    "\n",
    "For later..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_to_codes = {\"gorilla\": \"ggorilla\",\n",
    "            \"chimp\": \"ptroglodytes\",\n",
    "            \"marmoset\": \"cjacchus\",\n",
    "            \"rhesus\": \"mmulatta\",\n",
    "            \"human\": \"hsapiens\"\n",
    "             }\n",
    "\n",
    "## Locations of gtf files to map symbole to ensemble ID\n",
    "gtf_paths = {\"gorilla\": \"/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/great_apes/species/inputs/Gorilla_gorilla.gorGor4.110.gtf\",\n",
    "            \"chimp\": \"/allen/programs/celltypes/workgroups/rnaseqanalysis/references/chimp/ncbi/pantro/premrna/genes/genes.gtf\",\n",
    "            \"marmoset\": \"/allen/programs/celltypes/workgroups/rnaseqanalysis/references/marmoset/ncbi/mcalja1.2.pat.x/genome/genes/genes.gtf\",\n",
    "            \"rhesus\": \"/allen/programs/celltypes/workgroups/rnaseqanalysis/references/macaque/ncbi/mmul10/genome/genes/genes.gtf\",\n",
    "            \"human\": \"/allen/programs/celltypes/workgroups/hct/SEA-AD/RNAseq/cellxgene/input/genes.gtf\"}\n",
    "\n",
    "gtfs={}\n",
    "for s in gtf_paths.keys():\n",
    "    print(s)\n",
    "    gtfs[s]=gtfparse.parse_gtf_and_expand_attributes(gtf_paths[s])\n",
    "\n",
    "#Just take a peek\n",
    "for s in gtfs.keys():\n",
    "    print(s,species_to_codes[s])\n",
    "    print(gtfs[s])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load tables of unmapped genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_path='/home/matthew.schmitz/nhp_unmapped_genes'\n",
    "unmapped={}\n",
    "for p in os.listdir(gene_path):\n",
    "    if '.csv' in p:\n",
    "        s=p.split('_')[0]\n",
    "        unmapped[s]=pd.read_csv(os.path.join(gene_path,p))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "human\n",
      "50242\n",
      "chimp\n",
      "56936\n",
      "marmoset\n",
      "27125\n",
      "rhesus\n",
      "40278\n",
      "gorilla\n",
      "55217\n"
     ]
    }
   ],
   "source": [
    "for s in unmapped.keys():\n",
    "    print(s)\n",
    "    print(len(unmapped[s]['gene'].unique()))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fetch most current biomart tables\n",
    "\n",
    "Also appears that many missing IDs are symbols mixed in, match across several columns table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_biomart_table(dataset, filename):\n",
    "    #all_attributes = [x for x in dataset.attributes.keys() if 'hsapiens' in x]\n",
    "    all_attributes = [\n",
    "        'ensembl_gene_id',\n",
    "        'external_gene_name',\n",
    "        'hsapiens_homolog_associated_gene_name',\n",
    "        'hsapiens_homolog_ensembl_gene',\n",
    "        'hsapiens_homolog_orthology_confidence'\n",
    "    ]\n",
    "\n",
    "    response = dataset.search({\n",
    "        'attributes': all_attributes\n",
    "    }, header=1)  # header=1 will include the column names\n",
    "\n",
    "    with open(filename, 'wb') as f:\n",
    "        f.write('\\t'.join(all_attributes).encode('ascii')+b'\\n')\n",
    "        for line in response.iter_lines():\n",
    "            f.write(line + b'\\n')\n",
    "\n",
    "def find_matching_rows(strings, df,colnames=None):\n",
    "    matches = {}\n",
    "    if colnames is None:\n",
    "        colnames=df.columns\n",
    "    use_df=df.loc[:,colnames]\n",
    "    for s in tqdm.tqdm(strings):\n",
    "        mask = use_df.isin([s]).any(axis=1)\n",
    "        matches[s] = df[mask]\n",
    "    return matches\n",
    "\n",
    "def get_human_orthologs(species_name, identifiers, cache_path):\n",
    "    #Returns a dictionary of {original_key: dataframe of matched rows}\n",
    "    server = BiomartServer(\"http://www.ensembl.org/biomart\")\n",
    "    dataset_name = f\"{species_name.lower()}_gene_ensembl\"\n",
    "    \n",
    "    # Construct cache_path based on dataset_name\n",
    "    cache_name = os.path.join(cache_path,dataset_name + \"_table.txt\")\n",
    "    \n",
    "    dataset = server.datasets[dataset_name]\n",
    "\n",
    "    # Check cache and download if necessary\n",
    "    if not os.path.exists(cache_name):\n",
    "        print(\"Cache not found. Downloading BioMart table...\")\n",
    "        download_biomart_table(dataset, cache_name)\n",
    "        print(f\"Downloaded and saved to {cache_name}\")\n",
    "\n",
    "    # Read the table into a DataFrame\n",
    "    df = pd.read_csv(cache_name, sep='\\t', dtype=str)\n",
    "    \n",
    "    # Filter using Ensembl IDs and gene symbols and add 'original_identifier' column\n",
    "    #filtered_df = df[df['ensembl_gene_id'].isin(identifiers) | df['external_gene_name'].isin(identifiers)].copy()\n",
    "    #filtered_df['original_identifier'] = filtered_df.apply(lambda row: row['ensembl_gene_id'] if row['ensembl_gene_id'] in identifiers else row['external_gene_name'], axis=1)\n",
    "    #orthologs = {row['original_identifier']: row.drop('original_identifier').to_dict() for _, row in filtered_df.iterrows()}\n",
    "\n",
    "    orthologs=find_matching_rows(identifiers, df,df.columns[df.columns.str.contains('gene')])        \n",
    "    return orthologs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapped={}\n",
    "for s in unmapped.keys():\n",
    "    print(s)\n",
    "    if s=='human':\n",
    "        continue\n",
    "    mapped[s]=get_human_orthologs(species_to_codes[s],unmapped[s]['gene'].unique(),cache_path='/home/matthew.schmitz/cache/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concatenate_dataframes(dataframes,key_name='key'):\n",
    "    concatenated = pd.concat(dataframes.values(), keys=dataframes.keys(), axis=0).reset_index(level=1, drop=True).reset_index()\n",
    "    concatenated = concatenated.rename(columns={'index': key_name})\n",
    "    return concatenated\n",
    "\n",
    "mapped_dfs={}\n",
    "for s in mapped.keys():\n",
    "    mapped_dfs[s]=concatenate_dataframes(mapped[s],'original_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in mapped_dfs.keys():\n",
    "    mapped_dfs[s].to_csv(os.path.join('/home/matthew.schmitz/nhp_unmapped_genes',s+'_mapped.txt'),sep='\\t',header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56954, 3)\n",
      "(22267, 6)\n",
      "(27126, 3)\n",
      "(15255, 6)\n",
      "(40280, 3)\n",
      "(17005, 6)\n",
      "(55218, 3)\n",
      "(31768, 6)\n"
     ]
    }
   ],
   "source": [
    "for s in mapped.keys():\n",
    "    print(unmapped[s].shape)\n",
    "    print(mapped_dfs[s].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>hsapiens_homolog_associated_gene_name</th>\n",
       "      <th>hsapiens_homolog_ensembl_gene</th>\n",
       "      <th>hsapiens_homolog_orthology_confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>ACTL10</td>\n",
       "      <td>ENSPTRG00000044260</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACTL10</td>\n",
       "      <td>ENSG00000288649</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>ACTL10</td>\n",
       "      <td>ENSPTRG00000047541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACTL10</td>\n",
       "      <td>ENSG00000288649</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>ACTR2</td>\n",
       "      <td>ENSPTRG00000011990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>ENSG00000138071</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>ACTR2</td>\n",
       "      <td>ENSPTRG00000033914</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>ENSG00000138071</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>ADAP1</td>\n",
       "      <td>ENSPTRG00000041621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADAP1</td>\n",
       "      <td>ENSG00000105963</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22199</th>\n",
       "      <td>ZNF91</td>\n",
       "      <td>ENSPTRG00000043231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF91</td>\n",
       "      <td>ENSG00000167232</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22200</th>\n",
       "      <td>ENSPTRG00000019240</td>\n",
       "      <td>ENSPTRG00000019240</td>\n",
       "      <td>ZNF92</td>\n",
       "      <td>ZNF723</td>\n",
       "      <td>ENSG00000268696</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22201</th>\n",
       "      <td>ENSPTRG00000019240</td>\n",
       "      <td>ENSPTRG00000019240</td>\n",
       "      <td>ZNF92</td>\n",
       "      <td>ZNF737</td>\n",
       "      <td>ENSG00000237440</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22257</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>ENSPTRG00000021962</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZXDA</td>\n",
       "      <td>ENSG00000198205</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22258</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>ENSPTRG00000047457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZXDA</td>\n",
       "      <td>ENSG00000198205</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1275 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              original_id     ensembl_gene_id external_gene_name  \\\n",
       "203                ACTL10  ENSPTRG00000044260                NaN   \n",
       "204                ACTL10  ENSPTRG00000047541                NaN   \n",
       "218                 ACTR2  ENSPTRG00000011990                NaN   \n",
       "219                 ACTR2  ENSPTRG00000033914                NaN   \n",
       "287                 ADAP1  ENSPTRG00000041621                NaN   \n",
       "...                   ...                 ...                ...   \n",
       "22199               ZNF91  ENSPTRG00000043231                NaN   \n",
       "22200  ENSPTRG00000019240  ENSPTRG00000019240              ZNF92   \n",
       "22201  ENSPTRG00000019240  ENSPTRG00000019240              ZNF92   \n",
       "22257                ZXDA  ENSPTRG00000021962                NaN   \n",
       "22258                ZXDA  ENSPTRG00000047457                NaN   \n",
       "\n",
       "      hsapiens_homolog_associated_gene_name hsapiens_homolog_ensembl_gene  \\\n",
       "203                                  ACTL10               ENSG00000288649   \n",
       "204                                  ACTL10               ENSG00000288649   \n",
       "218                                   ACTR2               ENSG00000138071   \n",
       "219                                   ACTR2               ENSG00000138071   \n",
       "287                                   ADAP1               ENSG00000105963   \n",
       "...                                     ...                           ...   \n",
       "22199                                 ZNF91               ENSG00000167232   \n",
       "22200                                ZNF723               ENSG00000268696   \n",
       "22201                                ZNF737               ENSG00000237440   \n",
       "22257                                  ZXDA               ENSG00000198205   \n",
       "22258                                  ZXDA               ENSG00000198205   \n",
       "\n",
       "      hsapiens_homolog_orthology_confidence  \n",
       "203                                       1  \n",
       "204                                       1  \n",
       "218                                       1  \n",
       "219                                       0  \n",
       "287                                       1  \n",
       "...                                     ...  \n",
       "22199                                     0  \n",
       "22200                                     0  \n",
       "22201                                     0  \n",
       "22257                                     1  \n",
       "22258                                     1  \n",
       "\n",
       "[1275 rows x 6 columns]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Cases where one chimp symbol maps to multiple chimp ensids\n",
    "mapped_dfs['chimp'].loc[mapped_dfs['chimp']['original_id'].duplicated(keep=False),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>hsapiens_homolog_associated_gene_name</th>\n",
       "      <th>hsapiens_homolog_ensembl_gene</th>\n",
       "      <th>hsapiens_homolog_orthology_confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10717</th>\n",
       "      <td>NBPF12</td>\n",
       "      <td>ENSPTRG00000000232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NBPF12</td>\n",
       "      <td>ENSG00000268043</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10715</th>\n",
       "      <td>NBPF10</td>\n",
       "      <td>ENSPTRG00000000232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NBPF10</td>\n",
       "      <td>ENSG00000271425</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10714</th>\n",
       "      <td>NBPF1</td>\n",
       "      <td>ENSPTRG00000000232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NBPF1</td>\n",
       "      <td>ENSG00000219481</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10718</th>\n",
       "      <td>NBPF14</td>\n",
       "      <td>ENSPTRG00000000232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NBPF14</td>\n",
       "      <td>ENSG00000270629</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10728</th>\n",
       "      <td>NBPF9</td>\n",
       "      <td>ENSPTRG00000000232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NBPF9</td>\n",
       "      <td>ENSG00000269713</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20588</th>\n",
       "      <td>TSPY2</td>\n",
       "      <td>ENSPTRG00000052575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TSPY2</td>\n",
       "      <td>ENSG00000168757</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20582</th>\n",
       "      <td>TSPY1</td>\n",
       "      <td>ENSPTRG00000052575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TSPY1</td>\n",
       "      <td>ENSG00000258992</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12282</th>\n",
       "      <td>PGA3</td>\n",
       "      <td>ENSPTRG00000052811</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGA3</td>\n",
       "      <td>ENSG00000229859</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12288</th>\n",
       "      <td>PGA5</td>\n",
       "      <td>ENSPTRG00000052811</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGA5</td>\n",
       "      <td>ENSG00000256713</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12285</th>\n",
       "      <td>PGA4</td>\n",
       "      <td>ENSPTRG00000052811</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGA4</td>\n",
       "      <td>ENSG00000229183</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>731 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      original_id     ensembl_gene_id external_gene_name  \\\n",
       "10717      NBPF12  ENSPTRG00000000232                NaN   \n",
       "10715      NBPF10  ENSPTRG00000000232                NaN   \n",
       "10714       NBPF1  ENSPTRG00000000232                NaN   \n",
       "10718      NBPF14  ENSPTRG00000000232                NaN   \n",
       "10728       NBPF9  ENSPTRG00000000232                NaN   \n",
       "...           ...                 ...                ...   \n",
       "20588       TSPY2  ENSPTRG00000052575                NaN   \n",
       "20582       TSPY1  ENSPTRG00000052575                NaN   \n",
       "12282        PGA3  ENSPTRG00000052811                NaN   \n",
       "12288        PGA5  ENSPTRG00000052811                NaN   \n",
       "12285        PGA4  ENSPTRG00000052811                NaN   \n",
       "\n",
       "      hsapiens_homolog_associated_gene_name hsapiens_homolog_ensembl_gene  \\\n",
       "10717                                NBPF12               ENSG00000268043   \n",
       "10715                                NBPF10               ENSG00000271425   \n",
       "10714                                 NBPF1               ENSG00000219481   \n",
       "10718                                NBPF14               ENSG00000270629   \n",
       "10728                                 NBPF9               ENSG00000269713   \n",
       "...                                     ...                           ...   \n",
       "20588                                 TSPY2               ENSG00000168757   \n",
       "20582                                 TSPY1               ENSG00000258992   \n",
       "12282                                  PGA3               ENSG00000229859   \n",
       "12288                                  PGA5               ENSG00000256713   \n",
       "12285                                  PGA4               ENSG00000229183   \n",
       "\n",
       "      hsapiens_homolog_orthology_confidence  \n",
       "10717                                     0  \n",
       "10715                                     0  \n",
       "10714                                     0  \n",
       "10718                                     0  \n",
       "10728                                     0  \n",
       "...                                     ...  \n",
       "20588                                     0  \n",
       "20582                                     0  \n",
       "12282                                     0  \n",
       "12288                                     0  \n",
       "12285                                     0  \n",
       "\n",
       "[731 rows x 6 columns]"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#cases where multiple chimp symbols map to a single ensid\n",
    "mapped_dfs['chimp'].loc[mapped_dfs['chimp']['ensembl_gene_id'].duplicated(keep=False),:].sort_values('ensembl_gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gorilla ggorilla\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 30084/30084 [02:10<00:00, 230.29it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chimp ptroglodytes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 59133/59133 [03:56<00:00, 250.38it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "marmoset cjacchus\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 35750/35750 [03:37<00:00, 164.17it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rhesus mmulatta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 35219/35219 [04:16<00:00, 137.14it/s]\n"
     ]
    }
   ],
   "source": [
    "gtf_outs={}\n",
    "for s in gtfs.keys():\n",
    "    print(s,species_to_codes[s])\n",
    "    if s=='human':\n",
    "        continue\n",
    "    intermediate=get_human_orthologs(species_to_codes[s],gtfs[s]['gene_id'].unique(),cache_path='/home/matthew.schmitz/cache/')\n",
    "    gtf_outs[s]=concatenate_dataframes(intermediate,'original_id')\n",
    "    gtf_outs[s].to_csv(os.path.join('/home/matthew.schmitz/nhp_unmapped_genes',s+'_mapped_whole_GTF.txt'),sep='\\t',header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genomes",
   "language": "python",
   "name": "genomes"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
