{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "517b7262-2a73-4551-bd05-2da01df879fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "genomes_tab=pd.read_csv(os.path.expanduser('~/Reference_Genome_tracking.csv'),sep=',',dtype='str')\n",
    "genomes_tab['Species']=genomes_tab['Species'].str.replace(' ','_',regex=False)\n",
    "genomes_tab.index=genomes_tab['Species']\n",
    "genomes_tab=genomes_tab.loc[~genomes_tab.index.isna(),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d430530c-ee9b-47b4-82c1-2693f0663279",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9728: delphinus delphis\n",
      "103584: delphinus capensis\n",
      "9739: tursiops truncatus\n",
      "9735: stenella attenuata\n",
      "9736: stenella longirostris\n",
      "9737: stenella coeruleoalba\n",
      "90247: lagenorhynchus obliquidens\n",
      "46167: steno bredanensis\n",
      "82174: pseudoorca crassidens\n",
      "103596: peponocephala electra\n",
      "103592: feresa attenuata\n",
      "38241: globicephala macrorhynchus\n",
      "9731: globicephala melas\n",
      "83653: grampus griseus\n",
      "9733: orcinus orca\n",
      "103588: lissodelphis borealis\n",
      "9742: phocoena phocoena\n",
      "9744: phocoenoides dalli\n",
      "9760: ziphius cavirostris\n",
      "9755: physeter macrocephalus\n",
      "27615: kogia breviceps\n",
      "9752: kogia sima\n",
      "9771: balaenoptera musculus\n",
      "9767: balaenoptera acutorostrata\n",
      "9773: megaptera novaeangliae\n",
      "9764: eschrichtius robustus\n",
      "48708: mesoplodon densirostris\n",
      "9749: delphinapterus leucas\n",
      "40151: monodon monoceros\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "def get_taxid(species_name):\n",
    "    \"\"\"Get the taxid for a species name using NCBI's Entrez Utilities.\"\"\"\n",
    "    Entrez.email = \"your-email@example.com\"  # Provide your email address\n",
    "    handle = Entrez.esearch(db=\"taxonomy\", term=species_name)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    \n",
    "    # Return the first TaxID found\n",
    "    if record[\"IdList\"]:\n",
    "        return record[\"IdList\"][0]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "species_list = [\n",
    "    \"delphinus delphis\", \"delphinus capensis\", \"tursiops truncatus\", \n",
    "    \"stenella attenuata\", \"stenella longirostris\", \"stenella coeruleoalba\", \n",
    "    \"lagenorhynchus obliquidens\", \"steno bredanensis\", \"pseudoorca crassidens\", \n",
    "    \"peponocephala electra\", \"feresa attenuata\", \"globicephala macrorhynchus\", \n",
    "    \"globicephala melas\", \"grampus griseus\", \"orcinus orca\", \n",
    "    \"lissodelphis borealis\", \"phocoena phocoena\", \"phocoenoides dalli\", \n",
    "    \"ziphius cavirostris\", \"physeter macrocephalus\", \"kogia breviceps\", \n",
    "    \"kogia sima\", \"balaenoptera musculus\", \"balaenoptera acutorostrata\", \n",
    "    \"megaptera novaeangliae\", \"eschrichtius robustus\", \"mesoplodon densirostris\", \n",
    "    \"delphinapterus leucas\", \"monodon monoceros\"\n",
    "]\n",
    "\n",
    "# Fetch TaxIDs\n",
    "cetacean_dict = {get_taxid(species):species  for species in species_list}\n",
    "\n",
    "for species, taxid in cetacean_dict.items():\n",
    "    print(f\"{species}: {taxid}\")\n",
    "\n",
    "\n",
    "species_dict=dict(zip([9668,9669, 246437, 10116, 13616, 27679,39432, \n",
    "                     9823, 9361, 9986, 60711, 9598, \n",
    "                     30608, 10181, 37293, 9545, 9544, \n",
    "                     10090, 30611, 9685, 9595, 9606, \n",
    "                     9483, 42100, 9515, 9614, 9407, \n",
    "                     9555, 9999, 9615, 7955,9739,9742,37347,9913,9913,9593,2715852,9612,9720,9704,9785,9778,9813,9818,319819,142885,9258],\n",
    "                   [\"Wild Ferret\",\"Ferret\", \"Chinese.Treeshrew\", \"Rat\", \"Opossum\", \"Squirrel.monkey\",\"Squirrel.monkey.sp\", \n",
    "                     \"Pig\", \"Armadillo.Nine.banded\", \"Rabbit\", \"African.green.monkey\", \"Chimpanzee\", \n",
    "                     \"Mouse.lemur\", \"Naked.mole.rat\", \"Owl.monkey\", \"Macaque.pig.tailed\", \"Macaque.rhesus\", \n",
    "                     \"Mouse\", \"Galago\", \"Cat\", \"Gorilla\", \"Human\", \n",
    "                     \"Marmoset\", \"Vaquita\", \"Tufted.capuchin\", \"Coyote\", \"Egyptian.fruit.bat\", \n",
    "                     \"Olive.baboon\", \"Squirrel.arctic.ground\",\"Dog\",\"Zebrafish\",'Tursiops','Phocoena phocoena','Tupaia belangeri',\n",
    "                   \"Bos taurus\",'Gorilla gorilla', 'Cebus imitator','Wolf','harbor seal','sea lion','Afr savannah elephant','manatee','hyrax','aardvark','hh tenrec','blind mole rat','','platypus']))\n",
    "\n",
    "species_dict.update(cetacean_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3433f4ee-c642-4530-95a2-24385bb2e12e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{9668: 'Wild Ferret',\n",
       " 9669: 'Ferret',\n",
       " 246437: 'Chinese.Treeshrew',\n",
       " 10116: 'Rat',\n",
       " 13616: 'Opossum',\n",
       " 27679: 'Squirrel.monkey',\n",
       " 39432: 'Squirrel.monkey.sp',\n",
       " 9823: 'Pig',\n",
       " 9361: 'Armadillo.Nine.banded',\n",
       " 9986: 'Rabbit',\n",
       " 60711: 'African.green.monkey',\n",
       " 9598: 'Chimpanzee',\n",
       " 30608: 'Mouse.lemur',\n",
       " 10181: 'Naked.mole.rat',\n",
       " 37293: 'Owl.monkey',\n",
       " 9545: 'Macaque.pig.tailed',\n",
       " 9544: 'Macaque.rhesus',\n",
       " 10090: 'Mouse',\n",
       " 30611: 'Galago',\n",
       " 9685: 'Cat',\n",
       " 9595: 'Gorilla',\n",
       " 9606: 'Human',\n",
       " 9483: 'Marmoset',\n",
       " 42100: 'Vaquita',\n",
       " 9515: 'Tufted.capuchin',\n",
       " 9614: 'Coyote',\n",
       " 9407: 'Egyptian.fruit.bat',\n",
       " 9555: 'Olive.baboon',\n",
       " 9999: 'Squirrel.arctic.ground',\n",
       " 9615: 'Dog',\n",
       " 7955: 'Zebrafish',\n",
       " 9739: 'Tursiops',\n",
       " 9742: 'Phocoena phocoena',\n",
       " 37347: 'Tupaia belangeri',\n",
       " 9913: 'Gorilla gorilla',\n",
       " 9593: 'Cebus imitator',\n",
       " 2715852: 'Wolf',\n",
       " 9612: 'harbor seal',\n",
       " 9720: 'sea lion',\n",
       " 9704: 'Afr savannah elephant',\n",
       " 9785: 'manatee',\n",
       " 9778: 'hyrax',\n",
       " 9813: 'aardvark',\n",
       " 9818: 'hh tenrec',\n",
       " 319819: 'blind mole rat',\n",
       " 142885: '',\n",
       " 9258: 'platypus',\n",
       " '9728': 'delphinus delphis',\n",
       " '103584': 'delphinus capensis',\n",
       " '9739': 'tursiops truncatus',\n",
       " '9735': 'stenella attenuata',\n",
       " '9736': 'stenella longirostris',\n",
       " '9737': 'stenella coeruleoalba',\n",
       " '90247': 'lagenorhynchus obliquidens',\n",
       " '46167': 'steno bredanensis',\n",
       " '82174': 'pseudoorca crassidens',\n",
       " '103596': 'peponocephala electra',\n",
       " '103592': 'feresa attenuata',\n",
       " '38241': 'globicephala macrorhynchus',\n",
       " '9731': 'globicephala melas',\n",
       " '83653': 'grampus griseus',\n",
       " '9733': 'orcinus orca',\n",
       " '103588': 'lissodelphis borealis',\n",
       " '9742': 'phocoena phocoena',\n",
       " '9744': 'phocoenoides dalli',\n",
       " '9760': 'ziphius cavirostris',\n",
       " '9755': 'physeter macrocephalus',\n",
       " '27615': 'kogia breviceps',\n",
       " '9752': 'kogia sima',\n",
       " '9771': 'balaenoptera musculus',\n",
       " '9767': 'balaenoptera acutorostrata',\n",
       " '9773': 'megaptera novaeangliae',\n",
       " '9764': 'eschrichtius robustus',\n",
       " '48708': 'mesoplodon densirostris',\n",
       " '9749': 'delphinapterus leucas',\n",
       " '40151': 'monodon monoceros'}"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9124de0b-d9da-4431-aca7-486f00bb4ce0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No suitable assembly found for 9668\n",
      "No suitable assembly found for 9669\n",
      "Best assembly for 246437: GCA_033439345.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/033/439/345/GCA_033439345.1_KIZ_version_2/GCA_033439345.1_KIZ_version_2_assembly_report.txt\n",
      "No GTF\n",
      "Best assembly for 10116: GCA_021556685.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/556/685/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0_assembly_report.txt\n",
      "reference\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/015/227/675/GCF_015227675.2_mRatBN7.2/GCF_015227675.2_mRatBN7.2_assembly_report.txt\n",
      "Best assembly for 13616: GCF_027887165.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/027/887/165/GCF_027887165.1_mMonDom1.pri/GCF_027887165.1_mMonDom1.pri_assembly_report.txt\n",
      "No suitable assembly found for 27679\n",
      "No suitable assembly found for 39432\n",
      "Best assembly for 9823: GCA_030704935.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/704/935/GCA_030704935.1_ASM3070493v1/GCA_030704935.1_ASM3070493v1_assembly_report.txt\n",
      "Best assembly for 9361: GCF_030445035.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/030/445/035/GCF_030445035.1_mDasNov1.hap2/GCF_030445035.1_mDasNov1.hap2_assembly_report.txt\n",
      "Best assembly for 9986: GCF_009806435.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/806/435/GCF_009806435.1_UM_NZW_1.0/GCF_009806435.1_UM_NZW_1.0_assembly_report.txt\n",
      "Best assembly for 60711: GCF_000409795.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/409/795/GCF_000409795.2_Chlorocebus_sabeus_1.1/GCF_000409795.2_Chlorocebus_sabeus_1.1_assembly_report.txt\n",
      "No GTF\n",
      "Best assembly for 9598: GCA_028858805.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/858/805/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt_assembly_report.txt\n",
      "reference\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/880/755/GCF_002880755.1_Clint_PTRv2/GCF_002880755.1_Clint_PTRv2_assembly_report.txt\n",
      "Best assembly for 30608: GCF_000165445.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/165/445/GCF_000165445.2_Mmur_3.0/GCF_000165445.2_Mmur_3.0_assembly_report.txt\n",
      "Best assembly for 10181: GCA_944319715.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/319/715/GCA_944319715.1_Naked_mole-rat_maternal/GCA_944319715.1_Naked_mole-rat_maternal_assembly_report.txt\n",
      "No suitable assembly found for 37293\n",
      "No suitable assembly found for 9545\n",
      "No GTF\n",
      "Best assembly for 9544: GCF_003339765.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/003/339/765/GCF_003339765.1_Mmul_10/GCF_003339765.1_Mmul_10_assembly_report.txt\n",
      "reference\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/230/795/GCA_000230795.1_CR_1.0/GCA_000230795.1_CR_1.0_assembly_report.txt\n",
      "Best assembly for 10090: GCA_921998315.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/921/998/315/GCA_921998315.2_DBA_2J_v3/GCA_921998315.2_DBA_2J_v3_assembly_report.txt\n",
      "No suitable assembly found for 30611\n",
      "Best assembly for 9685: GCF_018350175.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/018/350/175/GCF_018350175.1_F.catus_Fca126_mat1.0/GCF_018350175.1_F.catus_Fca126_mat1.0_assembly_report.txt\n",
      "Best assembly for 9595: GCA_028885475.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/885/475/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat_assembly_report.txt\n",
      "Best assembly for 9606: GCA_011064465.2\n",
      "reference\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/011/064/465/GCA_011064465.2_Ash1_v2.2/GCA_011064465.2_Ash1_v2.2_assembly_report.txt\n",
      "Best assembly for 9483: GCF_009663435.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/663/435/GCF_009663435.1_Callithrix_jacchus_cj1700_1.1/GCF_009663435.1_Callithrix_jacchus_cj1700_1.1_assembly_report.txt\n",
      "Best assembly for 42100: GCF_008692025.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/008/692/025/GCF_008692025.1_mPhoSin1.pri/GCF_008692025.1_mPhoSin1.pri_assembly_report.txt\n",
      "No suitable assembly found for 9515\n",
      "No suitable assembly found for 9614\n",
      "No suitable assembly found for 9407\n",
      "Best assembly for 9555: GCA_008728515.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/008/728/515/GCA_008728515.2_Panubis1.1/GCA_008728515.2_Panubis1.1_assembly_report.txt\n",
      "No suitable assembly found for 9999\n",
      "Best assembly for 9615: GCA_013276365.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/013/276/365/GCA_013276365.2_UNSW_CanFamBas_1.2/GCA_013276365.2_UNSW_CanFamBas_1.2_assembly_report.txt\n",
      "Best assembly for 7955: GCA_033170195.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/033/170/195/GCA_033170195.1_ASM3317019v1/GCA_033170195.1_ASM3317019v1_assembly_report.txt\n",
      "Best assembly for 9739: GCF_011762595.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/011/762/595/GCF_011762595.1_mTurTru1.mat.Y/GCF_011762595.1_mTurTru1.mat.Y_assembly_report.txt\n",
      "No suitable assembly found for 9742\n",
      "No suitable assembly found for 37347\n",
      "No GTF\n",
      "Best assembly for 9913: GCA_947034695.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/947/034/695/GCA_947034695.1_seqoccin.Bt.char.v1.0/GCA_947034695.1_seqoccin.Bt.char.v1.0_assembly_report.txt\n",
      "reference\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/003/055/GCF_000003055.6_Bos_taurus_UMD_3.1.1/GCF_000003055.6_Bos_taurus_UMD_3.1.1_assembly_report.txt\n",
      "No suitable assembly found for 9593\n",
      "No suitable assembly found for 2715852\n",
      "Best assembly for 9612: GCA_905319855.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/905/319/855/GCA_905319855.2_mCanLor1.2/GCA_905319855.2_mCanLor1.2_assembly_report.txt\n",
      "https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/accession/GCA_905319855.2/download?include_annotation_type=GENOME_GTF,SEQUENCE_REPORT&filename=GCA_905319855.2.zip\n",
      "/home/matthew.schmitz/Matthew/genome/ncbi/9612-Canis_lupus/best\n",
      "No suitable assembly found for 9720\n",
      "Best assembly for 9704: GCF_009762305.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/762/305/GCF_009762305.2_mZalCal1.pri.v2/GCF_009762305.2_mZalCal1.pri.v2_assembly_report.txt\n",
      "https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/accession/GCF_009762305.2/download?include_annotation_type=GENOME_GTF,SEQUENCE_REPORT&filename=GCF_009762305.2.zip\n",
      "/home/matthew.schmitz/Matthew/genome/ncbi/9704-Zalophus_californianus/best\n",
      "Best assembly for 9785: GCA_030014295.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/014/295/GCA_030014295.1_mLoxAfr1.hap2/GCA_030014295.1_mLoxAfr1.hap2_assembly_report.txt\n",
      "https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/accession/GCA_030014295.1/download?include_annotation_type=GENOME_GTF,SEQUENCE_REPORT&filename=GCA_030014295.1.zip\n",
      "/home/matthew.schmitz/Matthew/genome/ncbi/9785-Loxodonta_africana/best\n",
      "No suitable assembly found for 9778\n",
      "No suitable assembly found for 9813\n",
      "No suitable assembly found for 9818\n",
      "No suitable assembly found for 319819\n",
      "No suitable assembly found for 142885\n",
      "Best assembly for 9728: GCA_030062865.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/062/865/GCA_030062865.1_ASM3006286v1/GCA_030062865.1_ASM3006286v1_assembly_report.txt\n",
      "No suitable assembly found for 103584\n",
      "Best assembly for 9739: GCF_011762595.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/011/762/595/GCF_011762595.1_mTurTru1.mat.Y/GCF_011762595.1_mTurTru1.mat.Y_assembly_report.txt\n",
      "No suitable assembly found for 9735\n",
      "No suitable assembly found for 9736\n",
      "Best assembly for 9737: GCA_951394435.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/951/394/435/GCA_951394435.1_mSteCoe1.1/GCA_951394435.1_mSteCoe1.1_assembly_report.txt\n",
      "No suitable assembly found for 90247\n",
      "Best assembly for 46167: GCA_028646385.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/646/385/GCA_028646385.1_Steno_bredanensis_HiC/GCA_028646385.1_Steno_bredanensis_HiC_assembly_report.txt\n",
      "No suitable assembly found for 82174\n",
      "No suitable assembly found for 103596\n",
      "No suitable assembly found for 103592\n",
      "No suitable assembly found for 38241\n",
      "Best assembly for 9731: GCF_963455315.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/963/455/315/GCF_963455315.1_mGloMel1.1/GCF_963455315.1_mGloMel1.1_assembly_report.txt\n",
      "Best assembly for 83653: GCA_028646425.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/646/425/GCA_028646425.1_Grampus_griseus_HiC/GCA_028646425.1_Grampus_griseus_HiC_assembly_report.txt\n",
      "Best assembly for 9733: GCF_937001465.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/937/001/465/GCF_937001465.1_mOrcOrc1.1/GCF_937001465.1_mOrcOrc1.1_assembly_report.txt\n",
      "No suitable assembly found for 103588\n",
      "No suitable assembly found for 9742\n",
      "No suitable assembly found for 9744\n",
      "No suitable assembly found for 9760\n",
      "Best assembly for 9755: GCF_002837175.3\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/837/175/GCF_002837175.3_ASM283717v5/GCF_002837175.3_ASM283717v5_assembly_report.txt\n",
      "Best assembly for 27615: GCF_026419965.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/026/419/965/GCF_026419965.1_mKogBre1_haplotype_1/GCF_026419965.1_mKogBre1_haplotype_1_assembly_report.txt\n",
      "No suitable assembly found for 9752\n",
      "Best assembly for 9771: GCF_009873245.2\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/873/245/GCF_009873245.2_mBalMus1.pri.v3/GCF_009873245.2_mBalMus1.pri.v3_assembly_report.txt\n",
      "Best assembly for 9767: GCF_949987535.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/949/987/535/GCF_949987535.1_mBalAcu1.1/GCF_949987535.1_mBalAcu1.1_assembly_report.txt\n",
      "No suitable assembly found for 9773\n",
      "Best assembly for 9764: GCA_028021215.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/021/215/GCA_028021215.1_mEscRob2.pri/GCA_028021215.1_mEscRob2.pri_assembly_report.txt\n",
      "Best assembly for 48708: GCF_025265405.1\n",
      "best\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/025/265/405/GCF_025265405.1_mMesDen1_primary_haplotype/GCF_025265405.1_mMesDen1_primary_haplotype_assembly_report.txt\n",
      "No suitable assembly found for 9749\n",
      "No suitable assembly found for 40151\n"
     ]
    }
   ],
   "source": [
    "#Download all ncbi assemblies\n",
    "\n",
    "from Bio import Entrez\n",
    "import os\n",
    "import requests\n",
    "from ftplib import FTP\n",
    "import re\n",
    "import ftplib\n",
    "import requests\n",
    "import zipfile\n",
    "import os\n",
    "import shutil\n",
    "import gzip\n",
    "Entrez.email = \"matthew.schmitz@alleninstitute.org\"\n",
    "\n",
    "def gzip_file(file_path):\n",
    "    \"\"\"Compresses a file using gzip.\"\"\"\n",
    "    with open(file_path, 'rb') as f_in:\n",
    "        with gzip.open(file_path + '.gz', 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "    os.remove(file_path)\n",
    "\n",
    "def download_file(url, filename):\n",
    "    response = requests.get(url, stream=True)\n",
    "    if response.status_code == 200:\n",
    "        with open(filename, 'wb') as f:\n",
    "            f.write(response.content)\n",
    "    else:\n",
    "        print(f\"Failed to download {url}\")\n",
    "\n",
    "def download_and_unzip(url, target_path, string_to_check):\n",
    "    # Download the zip file\n",
    "    response = requests.get(url)\n",
    "    zip_file_path = os.path.join(target_path, 'downloaded_file.zip')\n",
    "    with open(zip_file_path, 'wb') as f:\n",
    "        f.write(response.content)\n",
    "\n",
    "    # Unzip the file\n",
    "    with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:\n",
    "        zip_ref.extractall(target_path)\n",
    "\n",
    "    # Move all files from subdirectories to the target directory and gzip if needed\n",
    "    for root, dirs, files in os.walk(target_path):\n",
    "        for file in files:\n",
    "            file_path = os.path.join(root, file)\n",
    "            if file_path != zip_file_path:  # Avoid moving the zip file itself\n",
    "                # Check if the file needs to be gzipped\n",
    "                shutil.move(file_path, os.path.join(target_path, file))\n",
    "                if (string_to_check) in file and not file.endswith('.gz'):\n",
    "                    gzip_file(os.path.join(target_path, file))\n",
    "\n",
    "    # Remove the unzipped directory tree and the zip file\n",
    "    for root, dirs, files in os.walk(target_path, topdown=False):\n",
    "        for name in dirs:\n",
    "            shutil.rmtree(os.path.join(root, name))\n",
    "    os.remove(zip_file_path)\n",
    "\n",
    "def find_files_in_directory(directory, file_extension):\n",
    "    for file in os.listdir(directory):\n",
    "        if re.search(file_extension, file):\n",
    "            return file\n",
    "    return None\n",
    "\n",
    "\n",
    "\n",
    "def check_ftp_for_gtf(ftp_path):\n",
    "    ftp_url = \"ftp.ncbi.nlm.nih.gov\"\n",
    "    # Extract the relative path from the full FTP path\n",
    "    relative_ftp_path = '/'.join(ftp_path.split('/')[3:])  # Remove the 'ftp://ftp.ncbi.nlm.nih.gov/' part\n",
    "    \n",
    "    try:\n",
    "        with ftplib.FTP(ftp_url) as ftp:\n",
    "            ftp.login()\n",
    "            ftp.cwd(relative_ftp_path)\n",
    "            files = ftp.nlst()\n",
    "            for file in files:\n",
    "                if file.endswith('.gtf.gz'):\n",
    "                    return True\n",
    "            return False\n",
    "    except ftplib.all_errors as e:\n",
    "        print(f\"FTP error: {e}\")\n",
    "        return False\n",
    "\n",
    "def sanitize_filename(filename):\n",
    "    # Replace spaces with underscores\n",
    "    filename = filename.replace(\" \", \"_\")\n",
    "\n",
    "    # Replace special characters with dashes\n",
    "    # Define a regex pattern for unwanted characters\n",
    "    pattern = r'[^\\w\\-_.]'\n",
    "    filename = re.sub(pattern, '-', filename)\n",
    "\n",
    "    return filename\n",
    "\n",
    "def download_ftp_file(ftp_url, destination_filename):\n",
    "    # Parse the FTP URL\n",
    "    ftp_host, ftp_path = ftp_url.replace(\"ftp://\", \"\").split(\"/\", 1)\n",
    "\n",
    "    # Connect to the FTP server\n",
    "    with FTP(ftp_host) as ftp:\n",
    "        ftp.login()\n",
    "        \n",
    "        # Ensure the directory for the file exists\n",
    "        os.makedirs(os.path.dirname(destination_filename), exist_ok=True)\n",
    "\n",
    "        # Download the file\n",
    "        try:\n",
    "            with open(destination_filename, 'wb') as local_file:\n",
    "                ftp.retrbinary(f'RETR {ftp_path}', local_file.write)\n",
    "        except Exception as e:\n",
    "            print(f\"Error downloading {ftp_url}: {e}\")\n",
    "\n",
    "\n",
    "def get_best_assembly_and_download(species_name,overall_path='',redownload=False):\n",
    "    handle = Entrez.esummary(db=\"taxonomy\", id=species_name)\n",
    "    record = Entrez.read(handle)\n",
    "    scientific_name = record[0]['ScientificName']\n",
    "\n",
    "    handle = Entrez.esearch(db=\"assembly\", term=f\"{species_name}[Taxonomy ID]  AND (latest[filter] AND all[filter] NOT anomalous[filter]) AND (\\\"chromosome\\\"[Assembly Level] OR \\\"complete genome\\\"[Assembly Level] OR )\", retmax=10)\n",
    "    record = Entrez.read(handle)\n",
    "    best_assembly = None\n",
    "    best_assembly_gtf = None\n",
    "    max_contig_n50 = 0\n",
    "    max_contig_n50_ref = 0\n",
    "    for assembly_id in record['IdList']:\n",
    "        summary_handle = Entrez.esummary(db=\"assembly\", id=assembly_id)\n",
    "        summary = Entrez.read(summary_handle)\n",
    "        doc_summary = summary['DocumentSummarySet']['DocumentSummary'][0]\n",
    "        contig_n50 = int(doc_summary['ContigN50'])\n",
    "        if contig_n50 > max_contig_n50:\n",
    "            best_assembly = doc_summary\n",
    "            max_contig_n50 = contig_n50\n",
    "        if check_ftp_for_gtf(doc_summary['FtpPath_GenBank']) and (contig_n50 > max_contig_n50_ref):\n",
    "            best_assembly_gtf=doc_summary\n",
    "            max_contig_n50_ref = contig_n50\n",
    "\n",
    "    \n",
    "    if not best_assembly:\n",
    "        print(f\"No suitable assembly found for {species_name}\")\n",
    "        return\n",
    "\n",
    "    make_two=False\n",
    "    if best_assembly_gtf is not None:\n",
    "        if best_assembly['FtpPath_Assembly_rpt'] != best_assembly_gtf['FtpPath_Assembly_rpt']:\n",
    "            print('No GTF')\n",
    "            make_two=True\n",
    "    \n",
    "    print(f\"Best assembly for {species_name}: {best_assembly['AssemblyAccession']}\")\n",
    "\n",
    "\n",
    "    if make_two:\n",
    "        assemblies={'best':best_assembly,'reference':best_assembly_gtf}\n",
    "    elif best_assembly_gtf is None:\n",
    "        assemblies={'best':best_assembly}\n",
    "    else:\n",
    "        assemblies={'reference':best_assembly_gtf}\n",
    "        \n",
    "    for d in assemblies.keys():\n",
    "        print(d)\n",
    "        dirname=os.path.join(str(species_name)+\"-\"+sanitize_filename(scientific_name),d)\n",
    "        cur_assembly=assemblies[d]\n",
    "           \n",
    "        # Downloading the FASTA file\n",
    "        print(cur_assembly['FtpPath_Assembly_rpt'])\n",
    "        stat_ftp_url = cur_assembly['FtpPath_Assembly_rpt']\n",
    "        stat_name=sanitize_filename(os.path.basename(stat_ftp_url))#sanitize_filename(f\"{species_name}_assembly_report.txt\")\n",
    "        stat_destination = os.path.join(overall_path,dirname,stat_name)\n",
    "        download_ftp_file(stat_ftp_url, stat_destination)\n",
    "\n",
    "        fasta_ftp_url = re.sub('_assembly_report\\.txt','_genomic.fna.gz',cur_assembly['FtpPath_Assembly_rpt'])\n",
    "        fasta_name=sanitize_filename(os.path.basename(fasta_ftp_url))#sanitize_filename(f\"{species_name}_genomic.fna.gz\")\n",
    "        fasta_destination = os.path.join(overall_path,dirname,fasta_name)\n",
    "        if not redownload and os.path.exists(fasta_destination):\n",
    "            continue\n",
    "        download_ftp_file(fasta_ftp_url, fasta_destination)\n",
    "        if d != 'best':\n",
    "            gtf_ftp_url = re.sub('_assembly_report\\.txt','_genomic.gtf.gz',cur_assembly['FtpPath_Assembly_rpt'])\n",
    "            gtf_name=sanitize_filename(os.path.basename(gtf_ftp_url))#sanitize_filename(f\"{species_name}_genomic.gtf.gz\")\n",
    "            gtf_destination = os.path.join(overall_path,dirname,gtf_name)\n",
    "            #gtf_destination = f\"{species_name}/{species_name}_genomic.gtf.gz\"\n",
    "            download_ftp_file(gtf_ftp_url, gtf_destination)\n",
    "        else:\n",
    "            try:\n",
    "                url=f\"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/accession/{cur_assembly['AssemblyAccession']}/download?include_annotation_type=GENOME_GTF,SEQUENCE_REPORT&filename={cur_assembly['AssemblyAccession']}.zip\"\n",
    "                print(url)\n",
    "                print(os.path.join(overall_path,dirname))\n",
    "                download_and_unzip(url, os.path.join(overall_path,dirname),'.gtf')\n",
    "            except:\n",
    "                pass\n",
    "                \n",
    "\n",
    "species_list = species_dict.keys()\n",
    "#species_list=[9598]\n",
    "#species_list=[10116]\n",
    "for species in species_list:\n",
    "    try:\n",
    "        best_assembly=get_best_assembly_and_download(species,\n",
    "                                                     overall_path='/home/matthew.schmitz/Matthew/genome/ncbi',\n",
    "                                                    redownload=False)\n",
    "    except:\n",
    "        print(species, 'exploded')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "f6e6d840-e963-421d-a1fc-a988233bf47c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No suitable assembly found for 127582\n"
     ]
    }
   ],
   "source": [
    "best_assembly=get_best_assembly_and_download(127582,\n",
    "                                             overall_path='/home/matthew.schmitz/Matthew/genome/ncbi',\n",
    "                                            redownload=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "bc37dd8d-795e-4118-9638-76f0b29a6124",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.ensembl.org/info/genomes/mustela_putorius?content-type=application/json\n",
      "9668 failed\n",
      "https://rest.ensembl.org/info/genomes/mustela_putorius_furo?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/tupaia_chinensis?content-type=application/json\n",
      "246437 failed\n",
      "https://rest.ensembl.org/info/genomes/rattus_norvegicus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/monodelphis_domestica?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/saimiri_boliviensis?content-type=application/json\n",
      "27679 failed\n",
      "https://rest.ensembl.org/info/genomes/saimiri_boliviensis_boliviensis?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/sus_scrofa?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/dasypus_novemcinctus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/oryctolagus_cuniculus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/chlorocebus_sabaeus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/pan_troglodytes?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/microcebus_murinus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/heterocephalus_glaber?content-type=application/json\n",
      "10181 failed\n",
      "https://rest.ensembl.org/info/genomes/aotus_nancymaae?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/macaca_nemestrina?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/macaca_mulatta?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/mus_musculus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/otolemur_garnettii?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/felis_catus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/gorilla_gorilla_gorilla?content-type=application/json\n",
      "9595 failed\n",
      "https://rest.ensembl.org/info/genomes/homo_sapiens?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/callithrix_jacchus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/phocoena_sinus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/sapajus_apella?content-type=application/json\n",
      "9515 failed\n",
      "https://rest.ensembl.org/info/genomes/canis_latrans?content-type=application/json\n",
      "9614 failed\n",
      "https://rest.ensembl.org/info/genomes/rousettus_aegyptiacus?content-type=application/json\n",
      "9407 failed\n",
      "https://rest.ensembl.org/info/genomes/papio_anubis?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/urocitellus_parryii?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/canis_lupus_familiaris?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/danio_rerio?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/tursiops_truncatus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/phocoena_phocoena?content-type=application/json\n",
      "9742 failed\n",
      "https://rest.ensembl.org/info/genomes/tupaia_belangeri?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/bos_taurus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/gorilla_gorilla?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/cebus_imitator?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/canis_lupus?content-type=application/json\n",
      "9612 failed\n",
      "https://rest.ensembl.org/info/genomes/phoca_vitulina?content-type=application/json\n",
      "9720 failed\n",
      "https://rest.ensembl.org/info/genomes/zalophus_californianus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/loxodonta_africana?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/trichechus_manatus?content-type=application/json\n",
      "9778 failed\n",
      "https://rest.ensembl.org/info/genomes/procavia_capensis?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/orycteropus_afer?content-type=application/json\n",
      "9818 failed\n",
      "https://rest.ensembl.org/info/genomes/setifer_setosus?content-type=application/json\n",
      "319819 failed\n",
      "https://rest.ensembl.org/info/genomes/spalax_microphthalmus?content-type=application/json\n",
      "142885 failed\n",
      "https://rest.ensembl.org/info/genomes/delphinus_delphis?content-type=application/json\n",
      "9728 failed\n",
      "https://rest.ensembl.org/info/genomes/delphinus_capensis?content-type=application/json\n",
      "103584 failed\n",
      "https://rest.ensembl.org/info/genomes/tursiops_truncatus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/stenella_attenuata?content-type=application/json\n",
      "9735 failed\n",
      "https://rest.ensembl.org/info/genomes/stenella_longirostris?content-type=application/json\n",
      "9736 failed\n",
      "https://rest.ensembl.org/info/genomes/stenella_coeruleoalba?content-type=application/json\n",
      "9737 failed\n",
      "https://rest.ensembl.org/info/genomes/lagenorhynchus_obliquidens?content-type=application/json\n",
      "90247 failed\n",
      "https://rest.ensembl.org/info/genomes/steno_bredanensis?content-type=application/json\n",
      "46167 failed\n",
      "https://rest.ensembl.org/info/genomes/pseudorca_crassidens?content-type=application/json\n",
      "82174 failed\n",
      "https://rest.ensembl.org/info/genomes/peponocephala_electra?content-type=application/json\n",
      "103596 failed\n",
      "https://rest.ensembl.org/info/genomes/feresa_attenuata?content-type=application/json\n",
      "103592 failed\n",
      "https://rest.ensembl.org/info/genomes/globicephala_macrorhynchus?content-type=application/json\n",
      "38241 failed\n",
      "https://rest.ensembl.org/info/genomes/globicephala_melas?content-type=application/json\n",
      "9731 failed\n",
      "https://rest.ensembl.org/info/genomes/grampus_griseus?content-type=application/json\n",
      "83653 failed\n",
      "https://rest.ensembl.org/info/genomes/orcinus_orca?content-type=application/json\n",
      "9733 failed\n",
      "https://rest.ensembl.org/info/genomes/lissodelphis_borealis?content-type=application/json\n",
      "103588 failed\n",
      "https://rest.ensembl.org/info/genomes/phocoena_phocoena?content-type=application/json\n",
      "9742 failed\n",
      "https://rest.ensembl.org/info/genomes/phocoenoides_dalli?content-type=application/json\n",
      "9744 failed\n",
      "https://rest.ensembl.org/info/genomes/ziphius_cavirostris?content-type=application/json\n",
      "9760 failed\n",
      "https://rest.ensembl.org/info/genomes/physeter_catodon?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/kogia_breviceps?content-type=application/json\n",
      "27615 failed\n",
      "https://rest.ensembl.org/info/genomes/kogia_sima?content-type=application/json\n",
      "9752 failed\n",
      "https://rest.ensembl.org/info/genomes/balaenoptera_musculus?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/balaenoptera_acutorostrata?content-type=application/json\n",
      "9767 failed\n",
      "https://rest.ensembl.org/info/genomes/megaptera_novaeangliae?content-type=application/json\n",
      "9773 failed\n",
      "https://rest.ensembl.org/info/genomes/eschrichtius_robustus?content-type=application/json\n",
      "9764 failed\n",
      "https://rest.ensembl.org/info/genomes/mesoplodon_densirostris?content-type=application/json\n",
      "48708 failed\n",
      "https://rest.ensembl.org/info/genomes/delphinapterus_leucas?content-type=application/json\n",
      "https://rest.ensembl.org/info/genomes/monodon_monoceros?content-type=application/json\n"
     ]
    }
   ],
   "source": [
    "#Download all ensembl assemblies\n",
    "import requests\n",
    "import os\n",
    "import re\n",
    "from ftplib import FTP\n",
    "\n",
    "def sanitize_filename(filename):\n",
    "    # Replace spaces with underscores and special characters with dashes\n",
    "    filename = filename.replace(\" \", \"_\")\n",
    "    pattern = r'[^\\w\\-_.]'\n",
    "    return re.sub(pattern, '-', filename)\n",
    "\n",
    "def download_ftp_file(ftp_url, destination_filename):\n",
    "    # Parse and download from FTP URL\n",
    "    ftp_host, ftp_path = ftp_url.replace(\"ftp://\", \"\").split(\"/\", 1)\n",
    "    with FTP(ftp_host) as ftp:\n",
    "        ftp.login()\n",
    "        os.makedirs(os.path.dirname(destination_filename), exist_ok=True)\n",
    "        try:\n",
    "            with open(destination_filename, 'wb') as local_file:\n",
    "                ftp.retrbinary(f'RETR {ftp_path}', local_file.write)\n",
    "        except Exception as e:\n",
    "            print(f\"Error downloading {ftp_url}: {e}\")\n",
    "\n",
    "def get_species_name(tax_id):\n",
    "    \"\"\" Fetch species name using taxonomic ID from Ensembl \"\"\"\n",
    "    url = f\"https://rest.ensembl.org/taxonomy/id/{tax_id}?content-type=application/json\"\n",
    "    response = requests.get(url)\n",
    "    return response.json()['scientific_name']\n",
    "\n",
    "def find_matching_files(ftp, path, patterns):\n",
    "    \"\"\" Find files matching given patterns \"\"\"\n",
    "    ftp.cwd(path)\n",
    "    files = ftp.nlst()\n",
    "    matched_files = [file for file in files if any(pattern in file for pattern in patterns)]\n",
    "    return matched_files\n",
    "\n",
    "def download_ensembl_files(tax_id, path,redownload=False):\n",
    "    species_name = get_species_name(tax_id)\n",
    "    species_path = species_name.lower().replace(' ', '_')\n",
    "\n",
    "    # Fetch genome information from Ensembl\n",
    "    url = f\"https://rest.ensembl.org/info/genomes/{species_path}?content-type=application/json\"\n",
    "    print(url)\n",
    "    response = requests.get(url)\n",
    "    data = response.json()\n",
    "    assembly_name = data['assembly_name']\n",
    "\n",
    "    # Prepare directory name\n",
    "    dir_name = sanitize_filename(f\"{tax_id}-{species_name}\")\n",
    "    dir_path = os.path.join(path, dir_name)\n",
    "    os.makedirs(dir_path, exist_ok=True)\n",
    "\n",
    "    # Prepare FTP paths\n",
    "    ensembl_ftp_base = \"ftp.ensembl.org\"\n",
    "    fasta_ftp_dir = f\"/pub/release-110/fasta/{species_path}/dna/\"\n",
    "    gtf_ftp_dir = f\"/pub/release-110/gtf/{species_path}/\"\n",
    "\n",
    "    with FTP(ensembl_ftp_base) as ftp:\n",
    "        ftp.login()\n",
    "\n",
    "        # Find matching FASTA and GTF files\n",
    "        fa_patterns = [\".dna_sm.toplevel.fa.gz\"]\n",
    "        gtf_patterns = [\".gtf.gz\"]\n",
    "        unwanted_patterns = [\"chr\", \"hapl\", \"abinitio\"]\n",
    "\n",
    "        fasta_files = find_matching_files(ftp, fasta_ftp_dir, fa_patterns)\n",
    "        gtf_files = find_matching_files(ftp, gtf_ftp_dir, gtf_patterns)\n",
    "        fasta_files = [f for f in fasta_files if not any(p in f for p in unwanted_patterns)]\n",
    "        gtf_files = [f for f in gtf_files if not any(p in f for p in unwanted_patterns)]\n",
    "        if not redownload and os.path.exists(os.path.join(dir_path, fasta_files[0])):\n",
    "            return None\n",
    "        # Download files\n",
    "        if fasta_files:\n",
    "            download_ftp_file(f\"ftp://{ensembl_ftp_base}{fasta_ftp_dir}{fasta_files[0]}\", os.path.join(dir_path, fasta_files[0]))\n",
    "        if gtf_files:\n",
    "            download_ftp_file(f\"ftp://{ensembl_ftp_base}{gtf_ftp_dir}{gtf_files[0]}\", os.path.join(dir_path, gtf_files[0]))\n",
    "\n",
    "\n",
    "\n",
    "species_list = species_dict.keys()\n",
    "#species_list=[str(x) for x in species_list]\n",
    "for species in species_list:\n",
    "    #download_ensembl_files(species, '/home/matthew.schmitz/Matthew/genome/ensembl')\n",
    "    try:\n",
    "        download_ensembl_files(species, '/home/matthew.schmitz/Matthew/genome/ensembl',redownload=False)\n",
    "    except:\n",
    "        print(species,'failed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab421dca-cc5e-4c46-9ae0-1e5f5fe114b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sus_scrofa\n",
      "Sus_scrofa.Sscrofa11.1.dna_sm.toplevel.fa.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Polars found a filename. Ensure you pass a path to the file instead of a python file object when possible for best performance.\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'tag', 'projection_parent_transcript']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gorilla_gorilla_gorilla\n",
      "None\n",
      "Mus_musculus\n",
      "Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Polars found a filename. Ensure you pass a path to the file instead of a python file object when possible for best performance.\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Callithrix_jacchus\n",
      "Callithrix_jacchus.mCalJac1.pat.X.dna_sm.toplevel.fa.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Polars found a filename. Ensure you pass a path to the file instead of a python file object when possible for best performance.\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'projection_parent_transcript']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Danio_rerio\n",
      "Danio_rerio.GRCz11.dna_sm.toplevel.fa.gz\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import gtfparse\n",
    "import re\n",
    "import gzip\n",
    "\n",
    "def is_gzipped(file_path):\n",
    "    \"\"\" Check if the file is gzipped \"\"\"\n",
    "    with open(file_path, 'rb') as f:\n",
    "        return f.read(2) == b'\\x1f\\x8b'\n",
    "\n",
    "def calculate_N50(fasta_path):\n",
    "    file_open_func = gzip.open if is_gzipped(fasta_path) else open\n",
    "    with file_open_func(fasta_path, \"rt\") as handle:\n",
    "        lengths = [len(rec) for rec in SeqIO.parse(handle, \"fasta\")]\n",
    "    total_len = sum(lengths)\n",
    "    lengths.sort(reverse=True)\n",
    "    cumsum = 0\n",
    "    for length in lengths:\n",
    "        cumsum += length\n",
    "        if cumsum >= total_len / 2:\n",
    "            return length\n",
    "    return 0\n",
    "\n",
    "\"\"\"\n",
    "def count_genes_transcripts(gtf_path):\n",
    "    db = gffutils.create_db(gtf_path, dbfn=\":memory:\", force=True, keep_order=True, merge_strategy='merge', sort_attribute_values=True)\n",
    "    num_genes = len(set(db.features_of_type('gene_id')))\n",
    "    num_transcripts = len(set(db.features_of_type('transcript_id')))\n",
    "    return num_genes, num_transcripts\n",
    "\"\"\"\n",
    "\n",
    "def count_genes_transcripts(gtf_path):\n",
    "    try:\n",
    "        gtf_df = gtfparse.read_gtf(gtf_path)\n",
    "        num_genes = len(gtf_df['gene_id'].unique())\n",
    "        try:\n",
    "            num_transcripts = len(gtf_df['transcript_id'].unique())\n",
    "        except:\n",
    "            num_transcripts = num_genes\n",
    "        return num_genes, num_transcripts\n",
    "    except:\n",
    "        return 0,0\n",
    "\n",
    "def find_files_in_directory(directory, file_extension):\n",
    "    for file in os.listdir(directory):\n",
    "        if re.search(file_extension, file):\n",
    "            return file\n",
    "    return None\n",
    "\n",
    "def process_directory(dir_path,gtf=False):\n",
    "    for root, dirs, files in os.walk(dir_path):\n",
    "        for dir_name in dirs:\n",
    "            #try:\n",
    "            dir_full_path = os.path.join(root, dir_name)\n",
    "            #taxid, species_name = dir_name.split('-', 1)\n",
    "            try:\n",
    "                taxid, species_name =re.search('([0-9]+-[A-Za-z_]+)',dir_full_path).group(0).split('-',1)\n",
    "                common_name = species_name.replace(\"_\", \" \")\n",
    "            except:\n",
    "                continue\n",
    "            print(species_name)\n",
    "            # Find the FASTA and GTF files\n",
    "            fasta_file = find_files_in_directory(dir_full_path, r'\\.f(n)?a(\\.gz)?$')\n",
    "            print(fasta_file)\n",
    "            if not fasta_file:\n",
    "                continue\n",
    "\n",
    "            fasta_path = os.path.join(dir_full_path, fasta_file)\n",
    "\n",
    "            # Read assembly name for NCBI\n",
    "            assembly_name = fasta_file\n",
    "            assembly_report_path = os.path.join(dir_full_path, f\"{taxid}_assembly_report.txt\")\n",
    "            if os.path.exists(assembly_report_path):\n",
    "                with open(assembly_report_path) as f:\n",
    "                    for line in f:\n",
    "                        if 'Assembly name' in line:\n",
    "                            assembly_name = line.split(':')[1].strip()\n",
    "                            break\n",
    "                      \n",
    "            n50 = calculate_N50(fasta_path)\n",
    "            gtf_file = find_files_in_directory(dir_full_path, r'\\.gtf')\n",
    "            if gtf and (gtf_file is not None):\n",
    "                gtf_path = os.path.join(dir_full_path, gtf_file)\n",
    "                num_genes, num_transcripts = count_genes_transcripts(gtf_path)\n",
    "            else:\n",
    "                gtf_file = gtf_file\n",
    "                gtf_path = os.path.join(dir_full_path, gtf_file if gtf_file is not None else '')\n",
    "                num_genes = None\n",
    "                num_transcripts = None\n",
    "            \n",
    "            stat_dict = {\n",
    "                'Species': species_name,\n",
    "                'TaxID': taxid,\n",
    "                'Common Name': common_name,\n",
    "                'Genome Assembly Name': assembly_name,\n",
    "                'N50': n50,\n",
    "                'Number of Genes': num_genes,\n",
    "                'Number of Transcripts': num_transcripts,\n",
    "                'FASTA File': fasta_file,\n",
    "                'FASTA Path': fasta_path,\n",
    "                'GTF File': gtf_file,\n",
    "                'GTF Path': gtf_path\n",
    "            }\n",
    "            yield stat_dict\n",
    "            #except:\n",
    "            #    yield None\n",
    "\n",
    "ensembl_data = list(process_directory('/home/matthew.schmitz/Matthew/genome/ensembl',gtf=True))\n",
    "ncbi_data = list(process_directory('/home/matthew.schmitz/Matthew/genome/ncbi',gtf=True))\n",
    "other_data = list(process_directory('/home/matthew.schmitz/Matthew/genome/other',gtf=True))\n",
    "\n",
    "ensembl_data = [x for x in ensembl_data if x is not None]\n",
    "ncbi_data = [x for x in ncbi_data if x is not None]\n",
    "other_data = [x for x in other_data if x is not None]\n",
    "\n",
    "# Create DataFrames\n",
    "ensembl_df = pd.DataFrame(ensembl_data)\n",
    "ncbi_df = pd.DataFrame(ncbi_data)\n",
    "other_df = pd.DataFrame(other_data)\n",
    "\n",
    "# Combine DataFrames\n",
    "combined_df = pd.concat([ensembl_df, ncbi_df, other_df])\n",
    "print(combined_df)\n",
    "combined_df.to_csv('./AllGenomesData.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "e3237833-b71c-4adb-a57f-83b16cb64d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df = pd.read_csv('./AllGenomesData.csv',sep=',',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "b443c1d9-8c4d-425c-8ef6-6c563781c4ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.index=list(range(combined_df.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "b53f4e56-b1c3-4562-86e6-f485d9883d0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df=combined_df.loc[~combined_df['TaxID'].isna(),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "be645e85-24e8-4b6b-8639-40689cc1571b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>Common Name</th>\n",
       "      <th>Genome Assembly Name</th>\n",
       "      <th>N50</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcripts</th>\n",
       "      <th>FASTA File</th>\n",
       "      <th>FASTA Path</th>\n",
       "      <th>GTF File</th>\n",
       "      <th>GTF Path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sus_scrofa</td>\n",
       "      <td>9823</td>\n",
       "      <td>Sus scrofa</td>\n",
       "      <td>Sus_scrofa.Sscrofa11.1.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sus_scrofa.Sscrofa11.1.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/9...</td>\n",
       "      <td>Sus_scrofa.Sscrofa11.1.110.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/1...</td>\n",
       "      <td>Mus_musculus.GRCm39.110.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Callithrix_jacchus</td>\n",
       "      <td>9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>Callithrix_jacchus.mCalJac1.pat.X.dna_sm.tople...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Callithrix_jacchus.mCalJac1.pat.X.dna_sm.tople...</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/9...</td>\n",
       "      <td>Callithrix_jacchus.mCalJac1.pat.X.110.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Danio_rerio</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>Danio_rerio.GRCz11.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Danio_rerio.GRCz11.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/7...</td>\n",
       "      <td>Danio_rerio.GRCz11.110.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Otolemur_garnettii</td>\n",
       "      <td>30611</td>\n",
       "      <td>Otolemur garnettii</td>\n",
       "      <td>Otolemur_garnettii.OtoGar3.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Otolemur_garnettii.OtoGar3.dna_sm.toplevel.fa.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/3...</td>\n",
       "      <td>Otolemur_garnettii.OtoGar3.110.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/ensembl/3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Megaptera_novaeangliae</td>\n",
       "      <td>9773</td>\n",
       "      <td>Megaptera novaeangliae</td>\n",
       "      <td>GCA_004329385.1_megNov1_genomic.fixheaders.fna.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_004329385.1_megNov1_genomic.fixheaders.fna.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ncb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ncb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Trichechus_manatus</td>\n",
       "      <td>127582</td>\n",
       "      <td>Trichechus manatus</td>\n",
       "      <td>GCF_000243295.1_TriManLat1.0_genomic.fna.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_000243295.1_TriManLat1.0_genomic.fna.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ncb...</td>\n",
       "      <td>genomic.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ncb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Rousettus_aegyptiacus</td>\n",
       "      <td>9407</td>\n",
       "      <td>Rousettus aegyptiacus</td>\n",
       "      <td>GCF_001466805.2_Raegyp2.0_genomic.fna.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_001466805.2_Raegyp2.0_genomic.fna.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ncb...</td>\n",
       "      <td>genomic.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ncb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Tupaia_belangeri</td>\n",
       "      <td>37347</td>\n",
       "      <td>Tupaia belangeri</td>\n",
       "      <td>TS_3.0_genome.fa.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TS_3.0_genome.fa.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/tre...</td>\n",
       "      <td>TS_3.0.genomeannotation.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/tre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Macaca_mulatta</td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>GCF_003339765.1_Mmul_10_fixed.fa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCF_003339765.1_Mmul_10_fixed.fa</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ucs...</td>\n",
       "      <td>rhemac10_genes.gtf.gz</td>\n",
       "      <td>/home/matthew.schmitz/Matthew/genome/other/ucs...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Species   TaxID             Common Name  \\\n",
       "0               Sus_scrofa    9823              Sus scrofa   \n",
       "1             Mus_musculus   10090            Mus musculus   \n",
       "2       Callithrix_jacchus    9483      Callithrix jacchus   \n",
       "3              Danio_rerio    7955             Danio rerio   \n",
       "4       Otolemur_garnettii   30611      Otolemur garnettii   \n",
       "..                     ...     ...                     ...   \n",
       "77  Megaptera_novaeangliae    9773  Megaptera novaeangliae   \n",
       "78      Trichechus_manatus  127582      Trichechus manatus   \n",
       "79   Rousettus_aegyptiacus    9407   Rousettus aegyptiacus   \n",
       "80        Tupaia_belangeri   37347        Tupaia belangeri   \n",
       "81          Macaca_mulatta    9544          Macaca mulatta   \n",
       "\n",
       "                                 Genome Assembly Name  N50  Number of Genes  \\\n",
       "0        Sus_scrofa.Sscrofa11.1.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
       "1           Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
       "2   Callithrix_jacchus.mCalJac1.pat.X.dna_sm.tople...  NaN              NaN   \n",
       "3            Danio_rerio.GRCz11.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
       "4    Otolemur_garnettii.OtoGar3.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
       "..                                                ...  ...              ...   \n",
       "77  GCA_004329385.1_megNov1_genomic.fixheaders.fna.gz  NaN              NaN   \n",
       "78        GCF_000243295.1_TriManLat1.0_genomic.fna.gz  NaN              NaN   \n",
       "79           GCF_001466805.2_Raegyp2.0_genomic.fna.gz  NaN              NaN   \n",
       "80                                TS_3.0_genome.fa.gz  NaN              NaN   \n",
       "81                   GCF_003339765.1_Mmul_10_fixed.fa  NaN              NaN   \n",
       "\n",
       "    Number of Transcripts                                         FASTA File  \\\n",
       "0                     NaN       Sus_scrofa.Sscrofa11.1.dna_sm.toplevel.fa.gz   \n",
       "1                     NaN          Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz   \n",
       "2                     NaN  Callithrix_jacchus.mCalJac1.pat.X.dna_sm.tople...   \n",
       "3                     NaN           Danio_rerio.GRCz11.dna_sm.toplevel.fa.gz   \n",
       "4                     NaN   Otolemur_garnettii.OtoGar3.dna_sm.toplevel.fa.gz   \n",
       "..                    ...                                                ...   \n",
       "77                    NaN  GCA_004329385.1_megNov1_genomic.fixheaders.fna.gz   \n",
       "78                    NaN        GCF_000243295.1_TriManLat1.0_genomic.fna.gz   \n",
       "79                    NaN           GCF_001466805.2_Raegyp2.0_genomic.fna.gz   \n",
       "80                    NaN                                TS_3.0_genome.fa.gz   \n",
       "81                    NaN                   GCF_003339765.1_Mmul_10_fixed.fa   \n",
       "\n",
       "                                           FASTA Path  \\\n",
       "0   /home/matthew.schmitz/Matthew/genome/ensembl/9...   \n",
       "1   /home/matthew.schmitz/Matthew/genome/ensembl/1...   \n",
       "2   /home/matthew.schmitz/Matthew/genome/ensembl/9...   \n",
       "3   /home/matthew.schmitz/Matthew/genome/ensembl/7...   \n",
       "4   /home/matthew.schmitz/Matthew/genome/ensembl/3...   \n",
       "..                                                ...   \n",
       "77  /home/matthew.schmitz/Matthew/genome/other/ncb...   \n",
       "78  /home/matthew.schmitz/Matthew/genome/other/ncb...   \n",
       "79  /home/matthew.schmitz/Matthew/genome/other/ncb...   \n",
       "80  /home/matthew.schmitz/Matthew/genome/other/tre...   \n",
       "81  /home/matthew.schmitz/Matthew/genome/other/ucs...   \n",
       "\n",
       "                                        GTF File  \\\n",
       "0              Sus_scrofa.Sscrofa11.1.110.gtf.gz   \n",
       "1                 Mus_musculus.GRCm39.110.gtf.gz   \n",
       "2   Callithrix_jacchus.mCalJac1.pat.X.110.gtf.gz   \n",
       "3                  Danio_rerio.GRCz11.110.gtf.gz   \n",
       "4          Otolemur_garnettii.OtoGar3.110.gtf.gz   \n",
       "..                                           ...   \n",
       "77                                           NaN   \n",
       "78                                genomic.gtf.gz   \n",
       "79                                genomic.gtf.gz   \n",
       "80                TS_3.0.genomeannotation.gtf.gz   \n",
       "81                         rhemac10_genes.gtf.gz   \n",
       "\n",
       "                                             GTF Path  \n",
       "0   /home/matthew.schmitz/Matthew/genome/ensembl/9...  \n",
       "1   /home/matthew.schmitz/Matthew/genome/ensembl/1...  \n",
       "2   /home/matthew.schmitz/Matthew/genome/ensembl/9...  \n",
       "3   /home/matthew.schmitz/Matthew/genome/ensembl/7...  \n",
       "4   /home/matthew.schmitz/Matthew/genome/ensembl/3...  \n",
       "..                                                ...  \n",
       "77  /home/matthew.schmitz/Matthew/genome/other/ncb...  \n",
       "78  /home/matthew.schmitz/Matthew/genome/other/ncb...  \n",
       "79  /home/matthew.schmitz/Matthew/genome/other/ncb...  \n",
       "80  /home/matthew.schmitz/Matthew/genome/other/tre...  \n",
       "81  /home/matthew.schmitz/Matthew/genome/other/ucs...  \n",
       "\n",
       "[82 rows x 11 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.sort_values('N50')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "090aa117-4cc9-4d6a-a36e-71fbf712f85c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Sus_scrofa', 'Mus_musculus', 'Callithrix_jacchus', 'Danio_rerio', 'Otolemur_garnettii', 'Mustela_putorius_furo', 'Canis_lupus_familiaris', 'Procavia_capensis', 'Bos_taurus', 'Chlorocebus_sabaeus', 'Rattus_norvegicus', 'Macaca_nemestrina', 'Loxodonta_africana', 'Felis_catus', 'Zalophus_californianus', 'Delphinapterus_leucas', 'Balaenoptera_musculus', 'Cebus_imitator', 'Dasypus_novemcinctus', 'Monodelphis_domestica', 'Gorilla_gorilla', 'Physeter_catodon', 'Microcebus_murinus', 'Homo_sapiens', 'Saimiri_boliviensis_boliviensis', 'Macaca_mulatta', 'Urocitellus_parryii', 'Papio_anubis', 'Tupaia_belangeri', 'Aotus_nancymaae', 'Oryctolagus_cuniculus', 'Monodon_monoceros', 'Pan_troglodytes', 'Phocoena_sinus', 'Tursiops_truncatus', 'Stenella_coeruleoalba', 'Canis_lupus', 'Sus_scrofa', 'Gorilla_gorilla_gorilla', 'Mus_musculus', 'Callithrix_jacchus', 'Danio_rerio', 'Canis_lupus_familiaris', 'Bos_taurus', 'Bos_taurus', 'Chlorocebus_sabaeus', 'Eschrichtius_robustus', 'Balaenoptera_acutorostrata', 'Rattus_norvegicus', 'Rattus_norvegicus', 'Kogia_breviceps', 'Loxodonta_africana', 'Felis_catus', 'Ornithorhynchus_anatinus', 'Tupaia_chinensis', 'Globicephala_melas', 'Delphinus_delphis', 'Zalophus_californianus', 'Balaenoptera_musculus', 'Steno_bredanensis', 'Mesoplodon_densirostris', 'Heterocephalus_glaber', 'Dasypus_novemcinctus', 'Monodelphis_domestica', 'Physeter_catodon', 'Microcebus_murinus', 'Homo_sapiens', 'Macaca_mulatta', 'Macaca_mulatta', 'Papio_anubis', 'Oryctolagus_cuniculus', 'Pan_troglodytes', 'Pan_troglodytes', 'Grampus_griseus', 'Orcinus_orca', 'Phocoena_sinus', 'Tursiops_truncatus', 'Megaptera_novaeangliae', 'Trichechus_manatus', 'Rousettus_aegyptiacus', 'Tupaia_belangeri', 'Macaca_mulatta']\n",
      "\n",
      "                  /-Otolemur_garnettii\n",
      "               /-|\n",
      "              |   \\-Microcebus_murinus\n",
      "              |\n",
      "              |               /-Macaca_nemestrina\n",
      "              |            /-|\n",
      "              |         /-|   \\-Macaca_mulatta\n",
      "            /-|        |  |\n",
      "           |  |      /-|   \\-Papio_anubis\n",
      "           |  |     |  |\n",
      "           |  |   /-|   \\-Chlorocebus_sabaeus\n",
      "           |  |  |  |\n",
      "           |  |  |  |   /-Gorilla_gorilla\n",
      "           |  |  |   \\-|\n",
      "         /-|   \\-|     |   /-Homo_sapiens\n",
      "        |  |     |      \\-|\n",
      "        |  |     |         \\-Pan_troglodytes\n",
      "        |  |     |\n",
      "        |  |     |   /-Aotus_nancymaae\n",
      "        |  |      \\-|\n",
      "      /-|  |         \\-Callithrix_jacchus\n",
      "     |  |  |\n",
      "     |  |   \\-Tupaia_chinensis\n",
      "     |  |\n",
      "     |  |      /-Heterocephalus_glaber\n",
      "     |  |   /-|\n",
      "     |  |  |  |   /-Rattus_norvegicus\n",
      "     |   \\-|   \\-|\n",
      "     |     |      \\-Mus_musculus\n",
      "     |     |\n",
      "     |      \\-Oryctolagus_cuniculus\n",
      "     |\n",
      "     |                  /-Kogia_breviceps\n",
      "     |                 |\n",
      "     |               /-|      /-Tursiops_truncatus\n",
      "     |              |  |   /-|\n",
      "   /-|              |  |  |   \\-Orcinus_orca\n",
      "  |  |              |   \\-|\n",
      "  |  |            /-|     |   /-Delphinapterus_leucas\n",
      "  |  |           |  |      \\-|\n",
      "  |  |           |  |         \\-Monodon_monoceros\n",
      "  |  |           |  |\n",
      "  |  |         /-|  |   /-Eschrichtius_robustus\n",
      "  |  |        |  |   \\-|\n",
      "  |  |        |  |      \\-Balaenoptera_acutorostrata\n",
      "  |  |      /-|  |\n",
      "  |  |     |  |   \\-Bos_taurus\n",
      "  |  |     |  |\n",
      "  |  |     |   \\-Sus_scrofa\n",
      "--|  |   /-|\n",
      "  |  |  |  |         /-Canis_lupus\n",
      "  |  |  |  |      /-|\n",
      "  |  |  |  |   /-|   \\-Canis_lupus_familiaris\n",
      "  |   \\-|  |  |  |\n",
      "  |     |   \\-|   \\-Zalophus_californianus\n",
      "  |     |     |\n",
      "  |     |      \\-Felis_catus\n",
      "  |     |\n",
      "  |      \\-Rousettus_aegyptiacus\n",
      "  |\n",
      "  |   /-Dasypus_novemcinctus\n",
      "  |  |\n",
      "   \\-|      /-Trichechus_manatus\n",
      "     |   /-|\n",
      "      \\-|   \\-Procavia_capensis\n",
      "        |\n",
      "         \\-Loxodonta_africana\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from ete3 import Tree\n",
    "import copy\n",
    "\n",
    "def get_genera_from_species_list(species_list):\n",
    "    \"\"\"Extract genera from a list of species' scientific names.\"\"\"\n",
    "    return set(species.split()[0] for species in species_list)\n",
    "\n",
    "def prune_tree(newick_tree_path, genera):\n",
    "    \"\"\"Prune the tree to retain only species from the specified genera.\"\"\"\n",
    "    tree = Tree(newick_tree_path,format=1)\n",
    "    original_tree = copy.deepcopy(tree)\n",
    "    for leaf in tree.iter_leaves():\n",
    "        genus = leaf.name.split()[0]\n",
    "        if genus not in genera:\n",
    "            leaf.delete()\n",
    "    return tree,original_tree\n",
    "\n",
    "# Assuming DataFrame is loaded from a CSV or created from process_directory\n",
    "# df = pd.read_csv('your_dataframe.csv')  # Load your DataFrame here\n",
    "\n",
    "# Extract list of species and genera\n",
    "species_list = [x.replace(' ','_') for x in  combined_df['Species'].tolist()]\n",
    "print(species_list)\n",
    "genera = get_genera_from_species_list(species_list)\n",
    "\n",
    "# Path to your Newick tree\n",
    "newick_tree_path = '/home/matthew.schmitz/Matthew/genome/hal/241-mammalian-2020v2.phast-242.nh'\n",
    "\n",
    "# Prune the tree\n",
    "pruned_tree, original_tree = prune_tree(newick_tree_path, genera)\n",
    "\n",
    "# Optionally, save the pruned tree to a file\n",
    "pruned_tree.write(outfile=\"/home/matthew.schmitz/Matthew/genome/hal/pruned_tree.nwk\")\n",
    "\n",
    "# Display the pruned tree (optional)\n",
    "print(pruned_tree)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "a019ceec-6450-4636-9509-b21af6359d5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species not in the tree: ['Danio_rerio', 'Mustela_putorius_furo', 'Balaenoptera_musculus', 'Cebus_imitator', 'Monodelphis_domestica', 'Physeter_catodon', 'Saimiri_boliviensis_boliviensis', 'Urocitellus_parryii', 'Tupaia_belangeri', 'Phocoena_sinus', 'Stenella_coeruleoalba', 'Gorilla_gorilla_gorilla', 'Danio_rerio', 'Ornithorhynchus_anatinus', 'Globicephala_melas', 'Delphinus_delphis', 'Balaenoptera_musculus', 'Steno_bredanensis', 'Mesoplodon_densirostris', 'Monodelphis_domestica', 'Physeter_catodon', 'Grampus_griseus', 'Phocoena_sinus', 'Megaptera_novaeangliae', 'Tupaia_belangeri']\n"
     ]
    }
   ],
   "source": [
    "from ete3 import Tree\n",
    "\n",
    "def get_species_not_in_tree(species_list, tree):\n",
    "    # Extract species names from the tree\n",
    "    species_in_tree = set(leaf.name for leaf in tree.iter_leaves())\n",
    "\n",
    "    # Find species not in the tree\n",
    "    species_not_in_tree = [species for species in species_list if species not in species_in_tree]\n",
    "\n",
    "    return species_not_in_tree\n",
    "\n",
    "# Example usage\n",
    "species_list = [\n",
    "    \"delphinus delphis\", \"delphinus capensis\", \"tursiops truncatus\",\n",
    "    # ... other species ...\n",
    "    \"monodon monoceros\"\n",
    "]\n",
    "\n",
    "#newick_tree = \"(delphinus delphis, (tursiops truncatus, monodon monoceros));\"  # Replace with your actual Newick tree\n",
    "#tree = Tree(newick_tree)\n",
    "\n",
    "# Find species not in the tree\n",
    "species_not_in_tree = get_species_not_in_tree(combined_df['Species'], pruned_tree)\n",
    "print(\"Species not in the tree:\", species_not_in_tree)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "74ad26b9-1400-4d69-b1a6-147b171afcaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Danio_rerio is in the tree already!\n",
      "Tupaia_belangeri is in the tree already!\n",
      "Mustela_putorius_furo is in the tree already!\n",
      "Monodelphis_domestica is in the tree already!\n",
      "Urocitellus_parryii is in the tree already!\n",
      "Phocoena_sinus is in the tree already!\n",
      "Stenella_coeruleoalba is in the tree already!\n",
      "Globicephala_melas is in the tree already!\n",
      "Delphinus_delphis is in the tree already!\n",
      "Steno_bredanensis is in the tree already!\n",
      "Grampus_griseus is in the tree already!\n",
      "Physeter_catodon is in the tree already!\n",
      "Gorilla_gorilla_gorilla is in the tree already!\n"
     ]
    }
   ],
   "source": [
    "from ete3 import Tree\n",
    "\n",
    "def add_node_to_common_ancestor(tree, new_node_name, existing_node_names):\n",
    "    all_nodes = [node.name for node in tree.traverse()]\n",
    "    if new_node_name in all_nodes:\n",
    "        print(new_node_name,'is in the tree already!')\n",
    "        return None\n",
    "    # Find the common ancestor of the existing nodes\n",
    "    common_ancestor = tree.get_common_ancestor(existing_node_names)\n",
    "\n",
    "    # Calculate the mean branch length to the common ancestor\n",
    "    branch_lengths = [tree.get_distance(name,common_ancestor) for name in existing_node_names]\n",
    "    mean_branch_length = sum(branch_lengths) / len(branch_lengths)\n",
    "\n",
    "    # Create the new node and add it to the common ancestor\n",
    "    new_node = common_ancestor.add_child(name=new_node_name, dist=mean_branch_length)\n",
    "    \n",
    "    return new_node\n",
    "\n",
    "# Add the new node\n",
    "add_node_to_common_ancestor(pruned_tree, 'Danio_rerio', ['Dasypus_novemcinctus','Otolemur_garnettii'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Tupaia_belangeri', ['Tupaia_chinensis','Callithrix_jacchus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Mustela_putorius_furo', ['Canis_lupus_familiaris','Felis_catus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Monodelphis_domestica', ['Dasypus_novemcinctus','Otolemur_garnettii'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Urocitellus_parryii', ['Rattus_norvegicus','Mus_musculus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Phocoena_sinus', ['Orcinus_orca','Tursiops_truncatus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Stenella_coeruleoalba', ['Orcinus_orca','Tursiops_truncatus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Globicephala_melas', ['Orcinus_orca','Tursiops_truncatus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Delphinus_delphis', ['Orcinus_orca','Tursiops_truncatus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Steno_bredanensis', ['Orcinus_orca','Tursiops_truncatus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Grampus_griseus', ['Orcinus_orca','Tursiops_truncatus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Physeter_catodon', ['Orcinus_orca','Tursiops_truncatus'])\n",
    "add_node_to_common_ancestor(pruned_tree, 'Gorilla_gorilla_gorilla', ['Homo_sapiens','Pan_troglodytes'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "a18c2271-4e39-4ebc-8f73-07df8a8ad03e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fix_tree=Tree('((((((((((Otolemur_garnettii:0.0477624,Microcebus_murinus:0.0231249)1:0.0417511,(((((Macaca_nemestrina:0.00360292,Macaca_mulatta:0.002437)1:0.0042712,Papio_anubis:0.0062452)1:0.00406285,Chlorocebus_sabaeus:0.00521092)1:0.00497113,((((Homo_sapiens:0.00596652,(Pan_troglodytes,Pan_paniscus):0.00186328),Gorilla_gorilla_gorilla:0.0039149)1:0.00180399)1:0.01633, Pongo_abelii)),((Aotus_nancymaae,(Saimiri_boliviensis_boliviensis,(Cebus_imitator,Sapajus_apella))):0.0219195,Callithrix_jacchus:0.0215731)1:0.000848199)1:0.0781501)1:0.0179149,(Tupaia_chinensis:0.024326,Tupaia_belangeri:0.0714062))1:0.00425506,(((Spalax_microphthalmus,Heterocephalus_glaber):0.0546792,((Rattus_norvegicus:0.0874217,Mus_musculus:0.00924972),Urocitellus_parryii:0.0483357)1:0.0570242)1:0.0368586,Oryctolagus_cuniculus:0.0240056)1:0.0119995)1:0.0309181,((((((((((Tursiops_truncatus:0.00604089,Stenella_coeruleoalba:0.00575805,Delphinus_delphis:0.00575805),((((Globicephala_macrorhynchus,Globicephala_melas),Pseudorca_crassidens):0.00575805,Steno_bredanensis):0.00575805,Grampus_griseus:0.00575805)),Orcinus_orca:0.00547522),Phocoena_sinus:0.00575805),Mesoplodon_densirostris),((Kogia_breviceps,Kogia_sima),Physeter_catodon):0.00575805)1:0.00540204,(((Balaenoptera_musculus,Balaenoptera_acutorostrata),Megaptera_Novaeangliae),Eschrichtius_robustus)),Bos_taurus:0.00179196)1:0.0168069,Sus_scrofa:0.0546012)1:0.00922581,(Felis_catus:0.0044349,((((Canis_lupus_familiaris,Canis_lupus),Canis_latrans)):0.000719825,(Mustela_putorius_furo,(Phoca_vitulina,Zalophus_californianus))):0.00257736)1:0.054329,Rousettus_aegyptiacus)1:0.00390038)1:0.0186332,Dasypus_novemcinctus:0.0756033), ((Orycteropus_afer,Setifer_setosus),(Procavia_capensis,(Loxodonta_africana, Trichechus_manatus)))),Monodelphis_domestica:0.118419),Ornithorhynchus_anatinus),Danio_rerio:0.118419);')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "454f178a-75a1-47c4-8a45-ec5002f4dc39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                              /-Otolemur_garnettii\n",
      "                           /-|\n",
      "                          |   \\-Microcebus_murinus\n",
      "                          |\n",
      "                          |               /-Macaca_nemestrina\n",
      "                          |            /-|\n",
      "                          |         /-|   \\-Macaca_mulatta\n",
      "                          |        |  |\n",
      "                          |      /-|   \\-Papio_anubis\n",
      "                          |     |  |\n",
      "                        /-|     |   \\-Chlorocebus_sabaeus\n",
      "                       |  |     |\n",
      "                       |  |   /-|            /-Homo_sapiens\n",
      "                       |  |  |  |         /-|\n",
      "                       |  |  |  |        |  |   /-Pan_troglodytes\n",
      "                       |  |  |  |   /- /-|   \\-|\n",
      "                       |  |  |  |  |     |      \\-Pan_paniscus\n",
      "                       |  |  |   \\-|     |\n",
      "                       |  |  |     |      \\-Gorilla_gorilla_gorilla\n",
      "                       |   \\-|     |\n",
      "                       |     |      \\-Pongo_abelii\n",
      "                     /-|     |\n",
      "                    |  |     |      /-Aotus_nancymaae\n",
      "                    |  |     |   /-|\n",
      "                    |  |     |  |  |   /-Saimiri_boliviensis_boliviensis\n",
      "                    |  |     |  |   \\-|\n",
      "                    |  |      \\-|     |   /-Cebus_imitator\n",
      "                    |  |        |      \\-|\n",
      "                    |  |        |         \\-Sapajus_apella\n",
      "                    |  |        |\n",
      "                    |  |         \\-Callithrix_jacchus\n",
      "                  /-|  |\n",
      "                 |  |  |   /-Tupaia_chinensis\n",
      "                 |  |   \\-|\n",
      "                 |  |      \\-Tupaia_belangeri\n",
      "                 |  |\n",
      "                 |  |         /-Spalax_microphthalmus\n",
      "                 |  |      /-|\n",
      "                 |  |     |   \\-Heterocephalus_glaber\n",
      "                 |  |   /-|\n",
      "                 |  |  |  |      /-Rattus_norvegicus\n",
      "                 |  |  |  |   /-|\n",
      "                 |   \\-|   \\-|   \\-Mus_musculus\n",
      "                 |     |     |\n",
      "                 |     |      \\-Urocitellus_parryii\n",
      "                 |     |\n",
      "                 |      \\-Oryctolagus_cuniculus\n",
      "                 |\n",
      "                 |                              /-Tursiops_truncatus\n",
      "                 |                             |\n",
      "                 |                           /-|--Stenella_coeruleoalba\n",
      "                 |                          |  |\n",
      "                 |                          |   \\-Delphinus_delphis\n",
      "                 |                          |\n",
      "                 |                          |            /-Globicephala_macrorhynchus\n",
      "                 |                        /-|         /-|\n",
      "                 |                       |  |      /-|   \\-Globicephala_melas\n",
      "                 |                       |  |     |  |\n",
      "                 |                       |  |   /-|   \\-Pseudorca_crassidens\n",
      "                 |                     /-|  |  |  |\n",
      "                 |                    |  |   \\-|   \\-Steno_bredanensis\n",
      "               /-|                    |  |     |\n",
      "              |  |                  /-|  |      \\-Grampus_griseus\n",
      "              |  |                 |  |  |\n",
      "              |  |                 |  |   \\-Orcinus_orca\n",
      "              |  |               /-|  |\n",
      "              |  |              |  |   \\-Phocoena_sinus\n",
      "              |  |              |  |\n",
      "              |  |            /-|   \\-Mesoplodon_densirostris\n",
      "              |  |           |  |\n",
      "              |  |           |  |      /-Kogia_breviceps\n",
      "              |  |           |  |   /-|\n",
      "              |  |           |   \\-|   \\-Kogia_sima\n",
      "              |  |           |     |\n",
      "              |  |         /-|      \\-Physeter_catodon\n",
      "              |  |        |  |\n",
      "              |  |        |  |         /-Balaenoptera_musculus\n",
      "              |  |        |  |      /-|\n",
      "              |  |        |  |   /-|   \\-Balaenoptera_acutorostrata\n",
      "              |  |      /-|  |  |  |\n",
      "              |  |     |  |   \\-|   \\-Megaptera_Novaeangliae\n",
      "              |  |     |  |     |\n",
      "            /-|  |   /-|  |      \\-Eschrichtius_robustus\n",
      "           |  |  |  |  |  |\n",
      "           |  |  |  |  |   \\-Bos_taurus\n",
      "           |  |  |  |  |\n",
      "           |  |  |  |   \\-Sus_scrofa\n",
      "           |  |  |  |\n",
      "           |  |  |  |   /-Felis_catus\n",
      "           |  |  |  |  |\n",
      "           |  |  |  |  |            /-Canis_lupus_familiaris\n",
      "           |  |  |  |--|         /-|\n",
      "           |  |   \\-|  |   /- /-|   \\-Canis_lupus\n",
      "           |  |     |  |  |     |\n",
      "           |  |     |   \\-|      \\-Canis_latrans\n",
      "         /-|  |     |     |\n",
      "        |  |  |     |     |   /-Mustela_putorius_furo\n",
      "        |  |  |     |      \\-|\n",
      "        |  |  |     |        |   /-Phoca_vitulina\n",
      "        |  |  |     |         \\-|\n",
      "        |  |  |     |            \\-Zalophus_californianus\n",
      "        |  |  |     |\n",
      "        |  |  |      \\-Rousettus_aegyptiacus\n",
      "        |  |  |\n",
      "        |  |   \\-Dasypus_novemcinctus\n",
      "      /-|  |\n",
      "     |  |  |      /-Orycteropus_afer\n",
      "     |  |  |   /-|\n",
      "     |  |  |  |   \\-Setifer_setosus\n",
      "     |  |   \\-|\n",
      "     |  |     |   /-Procavia_capensis\n",
      "   /-|  |      \\-|\n",
      "  |  |  |        |   /-Loxodonta_africana\n",
      "  |  |  |         \\-|\n",
      "  |  |  |            \\-Trichechus_manatus\n",
      "--|  |  |\n",
      "  |  |   \\-Monodelphis_domestica\n",
      "  |  |\n",
      "  |   \\-Ornithorhynchus_anatinus\n",
      "  |\n",
      "   \\-Danio_rerio\n"
     ]
    }
   ],
   "source": [
    "print(fix_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "1676fec9-1f72-460d-add2-dd4ea609a1fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                              /-Otolemur_garnettii\n",
      "                           /-|\n",
      "                          |   \\-Microcebus_murinus\n",
      "                          |\n",
      "                          |               /-Macaca_nemestrina\n",
      "                          |            /-|\n",
      "                          |         /-|   \\-Macaca_mulatta\n",
      "                          |        |  |\n",
      "                        /-|      /-|   \\-Papio_anubis\n",
      "                       |  |     |  |\n",
      "                       |  |     |   \\-Chlorocebus_sabaeus\n",
      "                       |  |   /-|\n",
      "                       |  |  |  |         /-Homo_sapiens\n",
      "                       |  |  |  |      /-|\n",
      "                       |  |  |   \\- /-|   \\-Pan_troglodytes\n",
      "                       |  |  |        |\n",
      "                       |   \\-|         \\-Gorilla_gorilla_gorilla\n",
      "                     /-|     |\n",
      "                    |  |     |      /-Aotus_nancymaae\n",
      "                    |  |     |   /-|\n",
      "                    |  |     |  |  |   /-Saimiri_boliviensis_boliviensis\n",
      "                    |  |      \\-|   \\-|\n",
      "                    |  |        |      \\-Cebus_imitator\n",
      "                    |  |        |\n",
      "                    |  |         \\-Callithrix_jacchus\n",
      "                    |  |\n",
      "                  /-|  |   /-Tupaia_chinensis\n",
      "                 |  |   \\-|\n",
      "                 |  |      \\-Tupaia_belangeri\n",
      "                 |  |\n",
      "                 |  |            /-Rattus_norvegicus\n",
      "                 |  |         /-|\n",
      "                 |  |      /-|   \\-Mus_musculus\n",
      "                 |  |     |  |\n",
      "                 |  |   /-|   \\-Urocitellus_parryii\n",
      "                 |  |  |  |\n",
      "                 |   \\-|   \\-Heterocephalus_glaber\n",
      "                 |     |\n",
      "                 |      \\-Oryctolagus_cuniculus\n",
      "                 |\n",
      "                 |                              /-Tursiops_truncatus\n",
      "                 |                             |\n",
      "                 |                           /-|--Stenella_coeruleoalba\n",
      "                 |                          |  |\n",
      "                 |                          |   \\-Delphinus_delphis\n",
      "                 |                        /-|\n",
      "                 |                       |  |      /-Steno_bredanensis\n",
      "                 |                       |  |   /-|\n",
      "                 |                     /-|   \\-|   \\-Globicephala_melas\n",
      "                 |                    |  |     |\n",
      "               /-|                    |  |      \\-Grampus_griseus\n",
      "              |  |                  /-|  |\n",
      "              |  |                 |  |   \\-Orcinus_orca\n",
      "              |  |               /-|  |\n",
      "              |  |              |  |   \\-Phocoena_sinus\n",
      "              |  |              |  |\n",
      "              |  |            /-|   \\-Mesoplodon_densirostris\n",
      "              |  |           |  |\n",
      "              |  |           |  |   /-Physeter_catodon\n",
      "              |  |         /-|   \\-|\n",
      "              |  |        |  |      \\-Kogia_breviceps\n",
      "              |  |        |  |\n",
      "              |  |        |  |   /-Eschrichtius_robustus\n",
      "              |  |      /-|   \\-|\n",
      "              |  |     |  |     |   /-Balaenoptera_musculus\n",
      "              |  |     |  |      \\-|\n",
      "              |  |   /-|  |         \\-Balaenoptera_acutorostrata\n",
      "            /-|  |  |  |  |\n",
      "           |  |  |  |  |   \\-Bos_taurus\n",
      "           |  |  |  |  |\n",
      "           |  |  |  |   \\-Sus_scrofa\n",
      "           |  |  |  |\n",
      "           |  |  |  |   /-Felis_catus\n",
      "           |  |  |  |  |\n",
      "           |  |   \\-|--|         /-Canis_lupus_familiaris\n",
      "           |  |     |  |   /- /-|\n",
      "           |  |     |  |  |      \\-Canis_lupus\n",
      "         /-|  |     |   \\-|\n",
      "        |  |  |     |     |   /-Mustela_putorius_furo\n",
      "        |  |  |     |      \\-|\n",
      "        |  |  |     |         \\-Zalophus_californianus\n",
      "        |  |  |     |\n",
      "        |  |  |      \\-Rousettus_aegyptiacus\n",
      "        |  |  |\n",
      "      /-|  |   \\-Dasypus_novemcinctus\n",
      "     |  |  |\n",
      "     |  |  |   /-Procavia_capensis\n",
      "     |  |   \\-|\n",
      "     |  |     |   /-Loxodonta_africana\n",
      "   /-|  |      \\-|\n",
      "  |  |  |         \\-Trichechus_manatus\n",
      "  |  |  |\n",
      "--|  |   \\-Monodelphis_domestica\n",
      "  |  |\n",
      "  |   \\-Ornithorhynchus_anatinus\n",
      "  |\n",
      "   \\-Danio_rerio\n"
     ]
    }
   ],
   "source": [
    "species_list = [x.replace(' ','_') for x in  combined_df['Species'].tolist()]\n",
    "genera = get_genera_from_species_list(species_list)\n",
    "\n",
    "# Path to your Newick tree\n",
    "newick_tree_path = '/home/matthew.schmitz/Matthew/genome/hal/fixed_tree.nwk'\n",
    "\n",
    "# Prune the tree\n",
    "fixed_pruned_tree, original_tree = prune_tree(newick_tree_path, genera)\n",
    "\n",
    "# Optionally, save the pruned tree to a file\n",
    "fixed_pruned_tree.write(outfile=\"/home/matthew.schmitz/Matthew/genome/hal/fixed_pruned_tree.nwk\")\n",
    "\n",
    "print(fixed_pruned_tree) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "742b5313-d384-40be-a494-25857e0154ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n",
      "/scratch/fast/11984821/ipykernel_314158/1683586196.py:1: FutureWarning: The behavior of Series.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError\n",
      "  combined_df.groupby('Species')['N50'].idxmax()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Species\n",
       "Aotus_nancymaae                   NaN\n",
       "Balaenoptera_acutorostrata        NaN\n",
       "Balaenoptera_musculus             NaN\n",
       "Bos_taurus                        NaN\n",
       "Callithrix_jacchus                NaN\n",
       "Canis_lupus                       NaN\n",
       "Canis_lupus_familiaris            NaN\n",
       "Cebus_imitator                    NaN\n",
       "Chlorocebus_sabaeus               NaN\n",
       "Danio_rerio                       NaN\n",
       "Dasypus_novemcinctus              NaN\n",
       "Delphinapterus_leucas             NaN\n",
       "Delphinus_delphis                 NaN\n",
       "Eschrichtius_robustus             NaN\n",
       "Felis_catus                       NaN\n",
       "Globicephala_melas                NaN\n",
       "Gorilla_gorilla                   NaN\n",
       "Gorilla_gorilla_gorilla           NaN\n",
       "Grampus_griseus                   NaN\n",
       "Heterocephalus_glaber             NaN\n",
       "Homo_sapiens                      NaN\n",
       "Kogia_breviceps                   NaN\n",
       "Loxodonta_africana                NaN\n",
       "Macaca_mulatta                    NaN\n",
       "Macaca_nemestrina                 NaN\n",
       "Megaptera_novaeangliae            NaN\n",
       "Mesoplodon_densirostris           NaN\n",
       "Microcebus_murinus                NaN\n",
       "Monodelphis_domestica             NaN\n",
       "Monodon_monoceros                 NaN\n",
       "Mus_musculus                      NaN\n",
       "Mustela_putorius_furo             NaN\n",
       "Orcinus_orca                      NaN\n",
       "Ornithorhynchus_anatinus          NaN\n",
       "Oryctolagus_cuniculus             NaN\n",
       "Otolemur_garnettii                NaN\n",
       "Pan_troglodytes                   NaN\n",
       "Papio_anubis                      NaN\n",
       "Phocoena_sinus                    NaN\n",
       "Physeter_catodon                  NaN\n",
       "Procavia_capensis                 NaN\n",
       "Rattus_norvegicus                 NaN\n",
       "Rousettus_aegyptiacus             NaN\n",
       "Saimiri_boliviensis_boliviensis   NaN\n",
       "Stenella_coeruleoalba             NaN\n",
       "Steno_bredanensis                 NaN\n",
       "Sus_scrofa                        NaN\n",
       "Trichechus_manatus                NaN\n",
       "Tupaia_belangeri                  NaN\n",
       "Tupaia_chinensis                  NaN\n",
       "Tursiops_truncatus                NaN\n",
       "Urocitellus_parryii               NaN\n",
       "Zalophus_californianus            NaN\n",
       "Name: N50, dtype: float64"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df=combined_df.loc[combined_df.groupby('Species')['N50'].idxmax(),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "a91707c3-16e3-42c8-8486-1480d306354a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   Species   TaxID             Common Name  \\\n",
      "0               Sus_scrofa    9823              Sus scrofa   \n",
      "1             Mus_musculus   10090            Mus musculus   \n",
      "2       Callithrix_jacchus    9483      Callithrix jacchus   \n",
      "3              Danio_rerio    7955             Danio rerio   \n",
      "4       Otolemur_garnettii   30611      Otolemur garnettii   \n",
      "..                     ...     ...                     ...   \n",
      "0   Megaptera_novaeangliae    9773  Megaptera novaeangliae   \n",
      "1       Trichechus_manatus  127582      Trichechus manatus   \n",
      "2    Rousettus_aegyptiacus    9407   Rousettus aegyptiacus   \n",
      "3         Tupaia_belangeri   37347        Tupaia belangeri   \n",
      "4           Macaca_mulatta    9544          Macaca mulatta   \n",
      "\n",
      "                                 Genome Assembly Name  N50  Number of Genes  \\\n",
      "0        Sus_scrofa.Sscrofa11.1.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
      "1           Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
      "2   Callithrix_jacchus.mCalJac1.pat.X.dna_sm.tople...  NaN              NaN   \n",
      "3            Danio_rerio.GRCz11.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
      "4    Otolemur_garnettii.OtoGar3.dna_sm.toplevel.fa.gz  NaN              NaN   \n",
      "..                                                ...  ...              ...   \n",
      "0   GCA_004329385.1_megNov1_genomic.fixheaders.fna.gz  NaN              NaN   \n",
      "1         GCF_000243295.1_TriManLat1.0_genomic.fna.gz  NaN              NaN   \n",
      "2            GCF_001466805.2_Raegyp2.0_genomic.fna.gz  NaN              NaN   \n",
      "3                                 TS_3.0_genome.fa.gz  NaN              NaN   \n",
      "4                    GCF_003339765.1_Mmul_10_fixed.fa  NaN              NaN   \n",
      "\n",
      "    Number of Transcripts                                         FASTA File  \\\n",
      "0                     NaN       Sus_scrofa.Sscrofa11.1.dna_sm.toplevel.fa.gz   \n",
      "1                     NaN          Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz   \n",
      "2                     NaN  Callithrix_jacchus.mCalJac1.pat.X.dna_sm.tople...   \n",
      "3                     NaN           Danio_rerio.GRCz11.dna_sm.toplevel.fa.gz   \n",
      "4                     NaN   Otolemur_garnettii.OtoGar3.dna_sm.toplevel.fa.gz   \n",
      "..                    ...                                                ...   \n",
      "0                     NaN  GCA_004329385.1_megNov1_genomic.fixheaders.fna.gz   \n",
      "1                     NaN        GCF_000243295.1_TriManLat1.0_genomic.fna.gz   \n",
      "2                     NaN           GCF_001466805.2_Raegyp2.0_genomic.fna.gz   \n",
      "3                     NaN                                TS_3.0_genome.fa.gz   \n",
      "4                     NaN                   GCF_003339765.1_Mmul_10_fixed.fa   \n",
      "\n",
      "                                           FASTA Path  \\\n",
      "0   /home/matthew.schmitz/Matthew/genome/ensembl/9...   \n",
      "1   /home/matthew.schmitz/Matthew/genome/ensembl/1...   \n",
      "2   /home/matthew.schmitz/Matthew/genome/ensembl/9...   \n",
      "3   /home/matthew.schmitz/Matthew/genome/ensembl/7...   \n",
      "4   /home/matthew.schmitz/Matthew/genome/ensembl/3...   \n",
      "..                                                ...   \n",
      "0   /home/matthew.schmitz/Matthew/genome/other/ncb...   \n",
      "1   /home/matthew.schmitz/Matthew/genome/other/ncb...   \n",
      "2   /home/matthew.schmitz/Matthew/genome/other/ncb...   \n",
      "3   /home/matthew.schmitz/Matthew/genome/other/tre...   \n",
      "4   /home/matthew.schmitz/Matthew/genome/other/ucs...   \n",
      "\n",
      "                                        GTF File  \\\n",
      "0              Sus_scrofa.Sscrofa11.1.110.gtf.gz   \n",
      "1                 Mus_musculus.GRCm39.110.gtf.gz   \n",
      "2   Callithrix_jacchus.mCalJac1.pat.X.110.gtf.gz   \n",
      "3                  Danio_rerio.GRCz11.110.gtf.gz   \n",
      "4          Otolemur_garnettii.OtoGar3.110.gtf.gz   \n",
      "..                                           ...   \n",
      "0                                            NaN   \n",
      "1                                 genomic.gtf.gz   \n",
      "2                                 genomic.gtf.gz   \n",
      "3                 TS_3.0.genomeannotation.gtf.gz   \n",
      "4                          rhemac10_genes.gtf.gz   \n",
      "\n",
      "                                             GTF Path  \n",
      "0   /home/matthew.schmitz/Matthew/genome/ensembl/9...  \n",
      "1   /home/matthew.schmitz/Matthew/genome/ensembl/1...  \n",
      "2   /home/matthew.schmitz/Matthew/genome/ensembl/9...  \n",
      "3   /home/matthew.schmitz/Matthew/genome/ensembl/7...  \n",
      "4   /home/matthew.schmitz/Matthew/genome/ensembl/3...  \n",
      "..                                                ...  \n",
      "0   /home/matthew.schmitz/Matthew/genome/other/ncb...  \n",
      "1   /home/matthew.schmitz/Matthew/genome/other/ncb...  \n",
      "2   /home/matthew.schmitz/Matthew/genome/other/ncb...  \n",
      "3   /home/matthew.schmitz/Matthew/genome/other/tre...  \n",
      "4   /home/matthew.schmitz/Matthew/genome/other/ucs...  \n",
      "\n",
      "[82 rows x 11 columns]\n"
     ]
    }
   ],
   "source": [
    "print(combined_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "13fdab79-536c-4f99-8989-4291ec226161",
   "metadata": {},
   "outputs": [],
   "source": [
    " import pandas as pd\n",
    "\n",
    "def write_tree_and_dataframe_to_file(tree, df, filename, col1, col2):\n",
    "    \"\"\"\n",
    "    Writes a Newick tree, a space, and two DataFrame columns to a file.\n",
    "\n",
    "    :param tree: Newick tree string.\n",
    "    :param df: Pandas DataFrame.\n",
    "    :param filename: Output file name.\n",
    "    :param col1: First column name from DataFrame.\n",
    "    :param col2: Second column name from DataFrame.\n",
    "    \"\"\"\n",
    "    with open(filename, 'w') as f:\n",
    "        # Write the Newick tree\n",
    "        f.write(tree + '\\n\\n')  # Newick tree followed by a newline and a space\n",
    "\n",
    "        # Write the specified DataFrame columns\n",
    "        df[[col1, col2]].to_csv(f, sep='\\t', index=False,header=None)\n",
    "\n",
    "# Example usage\n",
    "newick_tree = fixed_pruned_tree.write()\n",
    "\n",
    "\n",
    "write_tree_and_dataframe_to_file(newick_tree, combined_df, '/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/genome/hal/26_plus_genomes/26pg.txt', 'Species', 'FASTA Path')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a0901a-e8ae-41ab-979b-9935427771f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genomes",
   "language": "python",
   "name": "genomes"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
