{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "517b7262-2a73-4551-bd05-2da01df879fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "genomes_tab=pd.read_csv(os.path.expanduser('~/Reference_Genome_tracking.csv'),sep=',',dtype='str')\n",
    "genomes_tab['Species']=genomes_tab['Species'].str.replace(' ','_',regex=False)\n",
    "genomes_tab.index=genomes_tab['Species']\n",
    "genomes_tab=genomes_tab.loc[~genomes_tab.index.isna(),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cb542cf6-7937-44f3-bc9c-37e15af0a42b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'221 Goodbye.'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from ftplib import FTP\n",
    "\n",
    "# Define the species of interest\n",
    "species_list = [\"homo_sapiens\", \"mus_musculus\", \"danio_rerio\"]\n",
    "\n",
    "ensembl_base = \"ftp.ensembl.org\"\n",
    "ensembl_path = \"/pub/current_gtf/\"\n",
    "\n",
    "ftp = FTP(ensembl_base)\n",
    "ftp.login()\n",
    "\n",
    "for species in species_list:\n",
    "    species_dir = ensembl_path + species + \"/\"\n",
    "    \n",
    "    # Get a list of files in the species directory\n",
    "    files = ftp.nlst(species_dir)\n",
    "    \n",
    "    # Identify the GTF file (assuming there's only one .gtf.gz per species directory)\n",
    "    gtf_file = next((f for f in files if f.endswith(\".gtf.gz\")), None)\n",
    "    \n",
    "    if gtf_file:\n",
    "        local_file_name = gtf_file.split('/')[-1]\n",
    "        with open(local_file_name, 'wb') as local_file:\n",
    "            ftp.retrbinary('RETR ' + gtf_file, local_file.write)\n",
    "\n",
    "ftp.quit()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "77e1ad30-580b-4932-87ec-f5a4acf7bb58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo sapiens\n",
      "{'Count': '1277', 'RetMax': '20', 'RetStart': '0', 'IdList': ['18601811', '16629201', '16629191', '16629181', '16629171', '16629161', '16629151', '16629131', '13464321', '12917181', '12656101', '12656051', '12656031', '12655981', '12655941', '12655831', '12655811', '12655741', '12655711', '12655631'], 'TranslationSet': [{'From': 'Homo sapiens', 'To': '\"Homo sapiens\"[Organism]'}], 'TranslationStack': [{'Term': '\"Homo sapiens\"[Organism]', 'Field': 'Organism', 'Count': '1277', 'Explode': 'Y'}, 'GROUP'], 'QueryTranslation': '\"Homo sapiens\"[Organism]'}\n",
      "Homo sapiens: DictElement({'RsUid': '', 'GbUid': '33120188', 'AssemblyAccession': 'GCA_913844075.1', 'LastMajorReleaseAccession': 'GCA_913844075.1', 'LatestAccession': '', 'ChainId': '913844075', 'AssemblyName': 's2r3_clone5_genome', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '9606', 'Organism': 'Homo sapiens (human)', 'SpeciesTaxid': '9606', 'SpeciesName': 'Homo sapiens', 'AssemblyType': 'haploid', 'AssemblyStatus': 'Scaffold', 'AssemblyStatusSort': '6', 'WGS': 'CAKNVN01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJEB28319', 'BioprojectId': '488786'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMEA8949014', 'BioSampleId': '21200168', 'Biosource': {'InfraspeciesList': [], 'Sex': '', 'Isolate': ''}, 'Coverage': '1', 'PartialGenomeRepresentation': 'false', 'Primary': '33120178', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2022/05/05 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2022/05/02 00:00', 'AsmUpdateDate': '2022/05/05 00:00', 'SubmissionDate': '2022/05/02 00:00', 'LastUpdateDate': '2022/05/05 00:00', 'SubmitterOrganization': 'UNIVERSITY OF LEEDS', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': ['genome length too large'], 'PropertyList': ['excluded-from-refseq', 'full-genome-representation', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_913844075.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '60974499', 'ScaffoldN50': '171058670', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/913/844/075/GCA_913844075.1_s2r3_clone5_genome', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/913/844/075/GCA_913844075.1_s2r3_clone5_genome/GCA_913844075.1_s2r3_clone5_genome_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/913/844/075/GCA_913844075.1_s2r3_clone5_genome/GCA_913844075.1_s2r3_clone5_genome_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '9C6X01019390255009138440759800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">1329</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">32</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">60974499</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">44</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">44</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">15</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">171058670</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">6120017525</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">5895405924</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/913/844/075/GCA_913844075.1_s2r3_clone5_genome/GCA_913844075.1_s2r3_clone5_genome_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/913/844/075/GCA_913844075.1_s2r3_clone5_genome</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/913/844/075/GCA_913844075.1_s2r3_clone5_genome/GCA_913844075.1_s2r3_clone5_genome_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>8</assembly-level> <assembly-status>Scaffold</assembly-status> <representative-status>na</representative-status> <submitter-organization>UNIVERSITY OF LEEDS</submitter-organization>    '}, attributes={'uid': '12655631'})\n",
      "Mus musculus\n",
      "{'Count': '97', 'RetMax': '20', 'RetStart': '0', 'IdList': ['18738731', '18738711', '16861651', '16568611', '16568401', '16199121', '16198941', '16189331', '16189291', '16189271', '16189241', '16189201', '16188821', '16188661', '16188531', '16188441', '16188431', '16188421', '16188411', '16188401'], 'TranslationSet': [{'From': 'Mus musculus', 'To': '\"Mus musculus\"[Organism]'}], 'TranslationStack': [{'Term': '\"Mus musculus\"[Organism]', 'Field': 'Organism', 'Count': '97', 'Explode': 'Y'}, 'GROUP'], 'QueryTranslation': '\"Mus musculus\"[Organism]'}\n",
      "Mus musculus: DictElement({'RsUid': '', 'GbUid': '41504078', 'AssemblyAccession': 'GCA_029234035.1', 'LastMajorReleaseAccession': 'GCA_029234035.1', 'LatestAccession': '', 'ChainId': '29234035', 'AssemblyName': 'ASM2923403v1', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '10090', 'Organism': 'Mus musculus (house mouse)', 'SpeciesTaxid': '10090', 'SpeciesName': 'Mus musculus', 'AssemblyType': 'haploid', 'AssemblyStatus': 'Scaffold', 'AssemblyStatusSort': '6', 'WGS': 'JARCGF01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJNA923323', 'BioprojectId': '923323'}], 'GB_Projects': [], 'RS_BioProjects': [], 'RS_Projects': [], 'BioSampleAccn': 'SAMN32816935', 'BioSampleId': '32816935', 'Biosource': {'InfraspeciesList': [{'Sub_type': 'strain', 'Sub_value': 'CC074'}], 'Sex': 'female', 'Isolate': ''}, 'Coverage': '100', 'PartialGenomeRepresentation': 'false', 'Primary': '41504068', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2023/03/17 00:00', 'AsmReleaseDate_RefSeq': '1/01/01 00:00', 'SeqReleaseDate': '2023/03/17 00:00', 'AsmUpdateDate': '2023/03/17 00:00', 'SubmissionDate': '2023/03/17 00:00', 'LastUpdateDate': '2023/03/17 00:00', 'SubmitterOrganization': 'The Jackson Laboratory for Genomic Medicine', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': [], 'PropertyList': ['full-genome-representation', 'latest', 'latest_genbank', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_029234035.1', 'RefSeq': '', 'Similarity': ''}, 'ContigN50': '3612783', 'ScaffoldN50': '3745907', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/029/234/035/GCA_029234035.1_ASM2923403v1', 'FtpPath_RefSeq': '', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/029/234/035/GCA_029234035.1_ASM2923403v1/GCA_029234035.1_ASM2923403v1_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/029/234/035/GCA_029234035.1_ASM2923403v1/GCA_029234035.1_ASM2923403v1_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '5C6X02029963872160292340359800', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">3796</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">206</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">3612783</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">3725</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">3725</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">197</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">3745907</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">2512845029</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">2512838174</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/029/234/035/GCA_029234035.1_ASM2923403v1/GCA_029234035.1_ASM2923403v1_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/029/234/035/GCA_029234035.1_ASM2923403v1</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/029/234/035/GCA_029234035.1_ASM2923403v1/GCA_029234035.1_ASM2923403v1_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>8</assembly-level> <assembly-status>Scaffold</assembly-status> <representative-status>na</representative-status> <submitter-organization>The Jackson Laboratory for Genomic Medicine</submitter-organization>    '}, attributes={'uid': '16188401'})\n",
      "Pan paniscus\n",
      "{'Count': '11', 'RetMax': '11', 'RetStart': '0', 'IdList': ['16840231', '16840221', '16214801', '16072021', '16072011', '9548001', '8641741', '8313331', '7133581', '474211', '396008'], 'TranslationSet': [{'From': 'Pan paniscus', 'To': '\"Pan paniscus\"[Organism]'}], 'TranslationStack': [{'Term': '\"Pan paniscus\"[Organism]', 'Field': 'Organism', 'Count': '11', 'Explode': 'Y'}, 'GROUP'], 'QueryTranslation': '\"Pan paniscus\"[Organism]'}\n",
      "Pan paniscus: DictElement({'RsUid': '396008', 'GbUid': '382348', 'AssemblyAccession': 'GCF_000258655.1', 'LastMajorReleaseAccession': 'GCF_000258655.1', 'LatestAccession': 'GCA_000258655.2', 'ChainId': '258655', 'AssemblyName': 'panpan1', 'UCSCName': '', 'EnsemblName': '', 'Taxid': '9597', 'Organism': 'Pan paniscus (pygmy chimpanzee)', 'SpeciesTaxid': '9597', 'SpeciesName': 'Pan paniscus', 'AssemblyType': 'haploid', 'AssemblyStatus': 'Scaffold', 'AssemblyStatusSort': '6', 'WGS': 'AJFE01', 'GB_BioProjects': [{'BioprojectAccn': 'PRJNA49285', 'BioprojectId': '49285'}], 'GB_Projects': [], 'RS_BioProjects': [{'BioprojectAccn': 'PRJNA169343', 'BioprojectId': '169343'}], 'RS_Projects': [], 'BioSampleAccn': 'SAMEA1029457', 'BioSampleId': '855949', 'Biosource': {'InfraspeciesList': [], 'Sex': 'female', 'Isolate': 'Ulindi'}, 'Coverage': '26', 'PartialGenomeRepresentation': 'false', 'Primary': '395948', 'AssemblyDescription': '', 'ReleaseLevel': 'Major', 'ReleaseType': 'Major', 'AsmReleaseDate_GenBank': '2012/05/28 00:00', 'AsmReleaseDate_RefSeq': '2012/07/05 00:00', 'SeqReleaseDate': '2012/04/24 00:00', 'AsmUpdateDate': '2012/09/13 00:00', 'SubmissionDate': '2012/04/24 00:00', 'LastUpdateDate': '2012/09/13 00:00', 'SubmitterOrganization': 'Max-Planck Institute for Evolutionary Anthropology', 'RefSeq_category': 'na', 'AnomalousList': [], 'ExclFromRefSeq': [], 'PropertyList': ['full-genome-representation', 'has-mitochondrion', 'has_annotation', 'refseq_has_annotation', 'replaced', 'replaced_genbank', 'suppressed_refseq', 'wgs'], 'FromType': '', 'Synonym': {'Genbank': 'GCA_000258655.1', 'RefSeq': 'GCF_000258655.1', 'Similarity': 'different'}, 'ContigN50': '66749', 'ScaffoldN50': '10124892', 'AnnotRptUrl': '', 'FtpPath_GenBank': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/258/655/GCA_000258655.1_panpan1', 'FtpPath_RefSeq': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/258/655/GCF_000258655.1_panpan1', 'FtpPath_Assembly_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/258/655/GCF_000258655.1_panpan1/GCF_000258655.1_panpan1_assembly_report.txt', 'FtpPath_Stats_rpt': 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/258/655/GCF_000258655.1_panpan1/GCF_000258655.1_panpan1_assembly_stats.txt', 'FtpPath_Regions_rpt': '', 'Busco': {'RefSeqAnnotationRelease': '', 'BuscoLineage': '', 'BuscoVer': '', 'Complete': '', 'SingleCopy': '', 'Duplicated': '', 'Fragmented': '', 'Missing': '', 'TotalCount': '0'}, 'SortOrder': '5C6X9C9A9999332500002586559898', 'Meta': ' <Stats> <Stat category=\"alt_loci_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"chromosome_count\" sequence_tag=\"all\">0</Stat> <Stat category=\"contig_count\" sequence_tag=\"all\">121253</Stat> <Stat category=\"contig_l50\" sequence_tag=\"all\">11032</Stat> <Stat category=\"contig_n50\" sequence_tag=\"all\">66749</Stat> <Stat category=\"non_chromosome_replicon_count\" sequence_tag=\"all\">1</Stat> <Stat category=\"replicon_count\" sequence_tag=\"all\">1</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"all\">10867</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"placed\">1</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unlocalized\">0</Stat> <Stat category=\"scaffold_count\" sequence_tag=\"unplaced\">10866</Stat> <Stat category=\"scaffold_l50\" sequence_tag=\"all\">73</Stat> <Stat category=\"scaffold_n50\" sequence_tag=\"all\">10124892</Stat> <Stat category=\"total_length\" sequence_tag=\"all\">2869190071</Stat> <Stat category=\"ungapped_length\" sequence_tag=\"all\">2725952766</Stat> </Stats> <FtpSites>   <FtpPath type=\"Assembly_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/258/655/GCF_000258655.1_panpan1/GCF_000258655.1_panpan1_assembly_report.txt</FtpPath>   <FtpPath type=\"GenBank\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/258/655/GCA_000258655.1_panpan1</FtpPath>   <FtpPath type=\"RefSeq\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/258/655/GCF_000258655.1_panpan1</FtpPath>   <FtpPath type=\"Stats_rpt\">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/258/655/GCF_000258655.1_panpan1/GCF_000258655.1_panpan1_assembly_stats.txt</FtpPath> </FtpSites> <assembly-level>2</assembly-level> <assembly-status>Scaffold</assembly-status> <representative-status>na</representative-status> <submitter-organization>Max-Planck Institute for Evolutionary Anthropology</submitter-organization>    '}, attributes={'uid': '396008'})\n"
     ]
    }
   ],
   "source": [
    "#Test function\n",
    "from Bio import Entrez\n",
    "from difflib import SequenceMatcher\n",
    "import re\n",
    "Entrez.email = \"matthew.schmitz@alleninstitute.org\"\n",
    "\n",
    "def get_ncbi_assembly_stats(species_name):\n",
    "    handle = Entrez.esearch(db=\"assembly\", term=species_name)\n",
    "    record = Entrez.read(handle)\n",
    "    print(record)\n",
    "    assembly_id = record['IdList'][-1]\n",
    "    summary_handle = Entrez.esummary(db=\"assembly\", id=assembly_id)\n",
    "    summary = Entrez.read(summary_handle)\n",
    "    \n",
    "    organism_from_ncbi = summary['DocumentSummarySet']['DocumentSummary'][0]['Organism']\n",
    "    match_ratio = SequenceMatcher(None, re.sub('\\(.+\\)','',organism_from_ncbi), species_name).ratio()\n",
    "\n",
    "    # Check if the match ratio is above a certain threshold\n",
    "    # You can adjust this threshold as needed\n",
    "    if match_ratio < 0.9:\n",
    "        print(f\"Warning: Organism from NCBI ('{organism_from_ncbi}') does not closely match the input species name ('{species_name}')!\")\n",
    "        return None\n",
    "\n",
    "    return summary['DocumentSummarySet']['DocumentSummary'][0]\n",
    "\n",
    "species_list = ['Homo sapiens', 'Mus musculus', 'Pan paniscus']\n",
    "for species in species_list:\n",
    "    print(species)\n",
    "    print(f\"{species}: {get_ncbi_assembly_stats(species)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9124de0b-d9da-4431-aca7-486f00bb4ce0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No suitable assembly found for 9668\n",
      "No suitable assembly found for 9669\n",
      "Best assembly for 246437: GCA_033439345.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/033/439/345/GCA_033439345.1_KIZ_version_2/GCA_033439345.1_KIZ_version_2_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/033/439/345/GCA_033439345.1_KIZ_version_2/GCA_033439345.1_KIZ_version_2_genomic.gtf.gz: 550 genomes/all/GCA/033/439/345/GCA_033439345.1_KIZ_version_2/GCA_033439345.1_KIZ_version_2_genomic.gtf.gz: No such file or directory\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/033/439/345/GCA_033439345.1_KIZ_version_2/GCA_033439345.1_KIZ_version_2_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/033/439/345/GCA_033439345.1_KIZ_version_2/GCA_033439345.1_KIZ_version_2_genomic.gtf.gz: 550 genomes/all/GCA/033/439/345/GCA_033439345.1_KIZ_version_2/GCA_033439345.1_KIZ_version_2_genomic.gtf.gz: No such file or directory\n",
      "Best assembly for 10116: GCA_021556685.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/556/685/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/556/685/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0_genomic.gtf.gz: 550 genomes/all/GCA/021/556/685/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0_genomic.gtf.gz: No such file or directory\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/556/685/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/021/556/685/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0_genomic.gtf.gz: 550 genomes/all/GCA/021/556/685/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0/GCA_021556685.1_UTH_Rnor_SHRSP_BbbUtx_1.0_genomic.gtf.gz: No such file or directory\n",
      "Best assembly for 13616: GCF_027887165.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/027/887/165/GCF_027887165.1_mMonDom1.pri/GCF_027887165.1_mMonDom1.pri_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/027/887/165/GCF_027887165.1_mMonDom1.pri/GCF_027887165.1_mMonDom1.pri_assembly_report.txt\n",
      "No suitable assembly found for 27679\n",
      "Best assembly for 9823: GCA_030704935.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/704/935/GCA_030704935.1_ASM3070493v1/GCA_030704935.1_ASM3070493v1_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/704/935/GCA_030704935.1_ASM3070493v1/GCA_030704935.1_ASM3070493v1_genomic.gtf.gz: 550 genomes/all/GCA/030/704/935/GCA_030704935.1_ASM3070493v1/GCA_030704935.1_ASM3070493v1_genomic.gtf.gz: No such file or directory\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/704/935/GCA_030704935.1_ASM3070493v1/GCA_030704935.1_ASM3070493v1_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/704/935/GCA_030704935.1_ASM3070493v1/GCA_030704935.1_ASM3070493v1_genomic.gtf.gz: 550 genomes/all/GCA/030/704/935/GCA_030704935.1_ASM3070493v1/GCA_030704935.1_ASM3070493v1_genomic.gtf.gz: No such file or directory\n",
      "Best assembly for 9361: GCF_030445035.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/030/445/035/GCF_030445035.1_mDasNov1.hap2/GCF_030445035.1_mDasNov1.hap2_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/030/445/035/GCF_030445035.1_mDasNov1.hap2/GCF_030445035.1_mDasNov1.hap2_assembly_report.txt\n",
      "Best assembly for 9986: GCF_009806435.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/806/435/GCF_009806435.1_UM_NZW_1.0/GCF_009806435.1_UM_NZW_1.0_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/806/435/GCF_009806435.1_UM_NZW_1.0/GCF_009806435.1_UM_NZW_1.0_assembly_report.txt\n",
      "Best assembly for 60711: GCF_000409795.2\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/409/795/GCF_000409795.2_Chlorocebus_sabeus_1.1/GCF_000409795.2_Chlorocebus_sabeus_1.1_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/409/795/GCF_000409795.2_Chlorocebus_sabeus_1.1/GCF_000409795.2_Chlorocebus_sabeus_1.1_assembly_report.txt\n",
      "Best assembly for 9598: GCA_028858805.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/858/805/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/858/805/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt_genomic.gtf.gz: 550 genomes/all/GCA/028/858/805/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt_genomic.gtf.gz: No such file or directory\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/858/805/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/858/805/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt_genomic.gtf.gz: 550 genomes/all/GCA/028/858/805/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt/GCA_028858805.1_NHGRI_mPanTro3-v1.1-hic.freeze_alt_genomic.gtf.gz: No such file or directory\n",
      "Best assembly for 30608: GCF_000165445.2\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/165/445/GCF_000165445.2_Mmur_3.0/GCF_000165445.2_Mmur_3.0_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/165/445/GCF_000165445.2_Mmur_3.0/GCF_000165445.2_Mmur_3.0_assembly_report.txt\n",
      "Best assembly for 10181: GCA_944319715.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/319/715/GCA_944319715.1_Naked_mole-rat_maternal/GCA_944319715.1_Naked_mole-rat_maternal_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/319/715/GCA_944319715.1_Naked_mole-rat_maternal/GCA_944319715.1_Naked_mole-rat_maternal_genomic.gtf.gz: 550 genomes/all/GCA/944/319/715/GCA_944319715.1_Naked_mole-rat_maternal/GCA_944319715.1_Naked_mole-rat_maternal_genomic.gtf.gz: No such file or directory\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/319/715/GCA_944319715.1_Naked_mole-rat_maternal/GCA_944319715.1_Naked_mole-rat_maternal_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/944/319/715/GCA_944319715.1_Naked_mole-rat_maternal/GCA_944319715.1_Naked_mole-rat_maternal_genomic.gtf.gz: 550 genomes/all/GCA/944/319/715/GCA_944319715.1_Naked_mole-rat_maternal/GCA_944319715.1_Naked_mole-rat_maternal_genomic.gtf.gz: No such file or directory\n",
      "No suitable assembly found for 37293\n",
      "No suitable assembly found for 9545\n",
      "Best assembly for 9544: GCF_003339765.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/003/339/765/GCF_003339765.1_Mmul_10/GCF_003339765.1_Mmul_10_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/003/339/765/GCF_003339765.1_Mmul_10/GCF_003339765.1_Mmul_10_assembly_report.txt\n",
      "Best assembly for 10090: GCA_921998315.2\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/921/998/315/GCA_921998315.2_DBA_2J_v3/GCA_921998315.2_DBA_2J_v3_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/921/998/315/GCA_921998315.2_DBA_2J_v3/GCA_921998315.2_DBA_2J_v3_genomic.gtf.gz: 550 genomes/all/GCA/921/998/315/GCA_921998315.2_DBA_2J_v3/GCA_921998315.2_DBA_2J_v3_genomic.gtf.gz: No such file or directory\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/921/998/315/GCA_921998315.2_DBA_2J_v3/GCA_921998315.2_DBA_2J_v3_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/921/998/315/GCA_921998315.2_DBA_2J_v3/GCA_921998315.2_DBA_2J_v3_genomic.gtf.gz: 550 genomes/all/GCA/921/998/315/GCA_921998315.2_DBA_2J_v3/GCA_921998315.2_DBA_2J_v3_genomic.gtf.gz: No such file or directory\n",
      "No suitable assembly found for 30611\n",
      "Best assembly for 9685: GCF_018350175.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/018/350/175/GCF_018350175.1_F.catus_Fca126_mat1.0/GCF_018350175.1_F.catus_Fca126_mat1.0_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/018/350/175/GCF_018350175.1_F.catus_Fca126_mat1.0/GCF_018350175.1_F.catus_Fca126_mat1.0_assembly_report.txt\n",
      "Best assembly for 9595: GCA_028885475.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/885/475/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/885/475/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat_genomic.gtf.gz: 550 genomes/all/GCA/028/885/475/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat_genomic.gtf.gz: No such file or directory\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/885/475/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat_assembly_report.txt\n",
      "Error downloading ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/885/475/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat_genomic.gtf.gz: 550 genomes/all/GCA/028/885/475/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat/GCA_028885475.1_NHGRI_mGorGor1-v1.1-0.2.freeze_pat_genomic.gtf.gz: No such file or directory\n",
      "Best assembly for 9606: GCA_011064465.2\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/011/064/465/GCA_011064465.2_Ash1_v2.2/GCA_011064465.2_Ash1_v2.2_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/011/064/465/GCA_011064465.2_Ash1_v2.2/GCA_011064465.2_Ash1_v2.2_assembly_report.txt\n",
      "Best assembly for 9483: GCF_009663435.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/663/435/GCF_009663435.1_Callithrix_jacchus_cj1700_1.1/GCF_009663435.1_Callithrix_jacchus_cj1700_1.1_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/663/435/GCF_009663435.1_Callithrix_jacchus_cj1700_1.1/GCF_009663435.1_Callithrix_jacchus_cj1700_1.1_assembly_report.txt\n",
      "Best assembly for 42100: GCF_008692025.1\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/008/692/025/GCF_008692025.1_mPhoSin1.pri/GCF_008692025.1_mPhoSin1.pri_assembly_report.txt\n",
      "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/008/692/025/GCF_008692025.1_mPhoSin1.pri/GCF_008692025.1_mPhoSin1.pri_assembly_report.txt\n"
     ]
    }
   ],
   "source": [
    "#Download all ncbi assemblies\n",
    "\n",
    "from Bio import Entrez\n",
    "import os\n",
    "import requests\n",
    "from ftplib import FTP\n",
    "import re\n",
    "\n",
    "Entrez.email = \"matthew.schmitz@alleninstitute.org\"\n",
    "\n",
    "\n",
    "def sanitize_filename(filename):\n",
    "    # Replace spaces with underscores\n",
    "    filename = filename.replace(\" \", \"_\")\n",
    "\n",
    "    # Replace special characters with dashes\n",
    "    # Define a regex pattern for unwanted characters\n",
    "    pattern = r'[^\\w\\-_.]'\n",
    "    filename = re.sub(pattern, '-', filename)\n",
    "\n",
    "    return filename\n",
    "\n",
    "def download_ftp_file(ftp_url, destination_filename):\n",
    "    # Parse the FTP URL\n",
    "    ftp_host, ftp_path = ftp_url.replace(\"ftp://\", \"\").split(\"/\", 1)\n",
    "\n",
    "    # Connect to the FTP server\n",
    "    with FTP(ftp_host) as ftp:\n",
    "        ftp.login()\n",
    "        \n",
    "        # Ensure the directory for the file exists\n",
    "        os.makedirs(os.path.dirname(destination_filename), exist_ok=True)\n",
    "\n",
    "        # Download the file\n",
    "        try:\n",
    "            with open(destination_filename, 'wb') as local_file:\n",
    "                ftp.retrbinary(f'RETR {ftp_path}', local_file.write)\n",
    "        except Exception as e:\n",
    "            print(f\"Error downloading {ftp_url}: {e}\")\n",
    "\n",
    "\n",
    "def get_best_assembly_and_download(species_name,overall_path=''):\n",
    "    handle = Entrez.esummary(db=\"taxonomy\", id=species_name)\n",
    "    record = Entrez.read(handle)\n",
    "    scientific_name = record[0]['ScientificName']\n",
    "\n",
    "    handle = Entrez.esearch(db=\"assembly\", term=f\"{species_name}[Taxonomy ID]  AND (latest[filter] AND all[filter] NOT anomalous[filter]) AND (\\\"chromosome\\\"[Assembly Level] OR \\\"complete genome\\\"[Assembly Level] OR )\", retmax=10)\n",
    "    record = Entrez.read(handle)\n",
    "    best_assembly = None\n",
    "    max_contig_n50 = 0\n",
    "    for assembly_id in record['IdList']:\n",
    "        summary_handle = Entrez.esummary(db=\"assembly\", id=assembly_id)\n",
    "        summary = Entrez.read(summary_handle)\n",
    "        doc_summary = summary['DocumentSummarySet']['DocumentSummary'][0]\n",
    "        contig_n50 = int(doc_summary['ContigN50'])\n",
    "        if contig_n50 > max_contig_n50:\n",
    "            best_assembly = doc_summary\n",
    "            max_contig_n50 = contig_n50\n",
    "\n",
    "    if not best_assembly:\n",
    "        print(f\"No suitable assembly found for {species_name}\")\n",
    "        return\n",
    "\n",
    "    print(f\"Best assembly for {species_name}: {best_assembly['AssemblyAccession']}\")\n",
    "\n",
    "    dirname=str(species_name)+\"-\"+sanitize_filename(scientific_name)\n",
    "    \n",
    "    # Downloading the FASTA file\n",
    "    for prefix in ['GCA', 'GCF']:\n",
    "        print(best_assembly['FtpPath_Assembly_rpt'])\n",
    "        stat_ftp_url = best_assembly['FtpPath_Assembly_rpt']\n",
    "        stat_name=sanitize_filename(f\"{species_name}_assembly_report.txt\")\n",
    "        stat_destination = os.path.join(overall_path,dirname,stat_name)\n",
    "        download_ftp_file(stat_ftp_url, stat_destination)\n",
    "        \n",
    "        fasta_ftp_url = re.sub('_assembly_report\\.txt','_genomic.fna.gz',best_assembly['FtpPath_Assembly_rpt'])\n",
    "        fasta_name=sanitize_filename(f\"{species_name}_genomic.fna.gz\")\n",
    "        fasta_destination = os.path.join(overall_path,dirname,fasta_name)\n",
    "        download_ftp_file(fasta_ftp_url, fasta_destination)\n",
    "\n",
    "        gtf_ftp_url = re.sub('_assembly_report\\.txt','_genomic.gtf.gz',best_assembly['FtpPath_Assembly_rpt'])\n",
    "        gtf_name=sanitize_filename(f\"{species_name}_genomic.gtf.gz\")\n",
    "        gtf_destination = os.path.join(overall_path,dirname,gtf_name)\n",
    "        #gtf_destination = f\"{species_name}/{species_name}_genomic.gtf.gz\"\n",
    "        download_ftp_file(gtf_ftp_url, gtf_destination)\n",
    "\n",
    "def download_file(url, filename):\n",
    "    response = requests.get(url, stream=True)\n",
    "    if response.status_code == 200:\n",
    "        with open(filename, 'wb') as f:\n",
    "            f.write(response.content)\n",
    "    else:\n",
    "        print(f\"Failed to download {url}\")\n",
    "\n",
    "\n",
    "species_dict=dict(zip([9668,9669, 246437, 10116, 13616, 27679, \n",
    "                     9823, 9361, 9986, 60711, 9598, \n",
    "                     30608, 10181, 37293, 9545, 9544, \n",
    "                     10090, 30611, 9685, 9595, 9606, \n",
    "                     9483, 42100, 9515, 9614, 9407, \n",
    "                     9555, 9999, 9615, 7955,9739,9742,37347,9913,9913,9593],\n",
    "                   [\"Wild Ferret\",\"Ferret\", \"Chinese.Treeshrew\", \"Rat\", \"Opossum\", \"Squirrel.monkey\", \n",
    "                     \"Pig\", \"Armadillo.Nine.banded\", \"Rabbit\", \"African.green.monkey\", \"Chimpanzee\", \n",
    "                     \"Mouse.lemur\", \"Naked.mole.rat\", \"Owl.monkey\", \"Macaque.pig.tailed\", \"Macaque.rhesus\", \n",
    "                     \"Mouse\", \"Galago\", \"Cat\", \"Gorilla\", \"Human\", \n",
    "                     \"Marmoset\", \"Vaquita\", \"Tufted.capuchin\", \"Coyote\", \"Egyptian.fruit.bat\", \n",
    "                     \"Olive.baboon\", \"Squirrel.arctic.ground\",\"Dog\",\"Zebrafish\",'Tursiops','Phocoena phocoena','Tupaia belangeri',\n",
    "                   \"Bos taurus\",'Gorilla gorilla']))\n",
    "\n",
    "species_list = species_dict.keys()\n",
    "#species_list=[str(x) for x in species_list]\n",
    "for species in species_list:\n",
    "    best_assembly=get_best_assembly_and_download(species,overall_path='/home/matthew.schmitz/Matthew/genome/ncbi')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc37dd8d-795e-4118-9638-76f0b29a6124",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import os\n",
    "import re\n",
    "from ftplib import FTP\n",
    "\n",
    "def sanitize_filename(filename):\n",
    "    # Replace spaces with underscores and special characters with dashes\n",
    "    filename = filename.replace(\" \", \"_\")\n",
    "    pattern = r'[^\\w\\-_.]'\n",
    "    return re.sub(pattern, '-', filename)\n",
    "\n",
    "def download_ftp_file(ftp_url, destination_filename):\n",
    "    # Parse and download from FTP URL\n",
    "    ftp_host, ftp_path = ftp_url.replace(\"ftp://\", \"\").split(\"/\", 1)\n",
    "    with FTP(ftp_host) as ftp:\n",
    "        ftp.login()\n",
    "        os.makedirs(os.path.dirname(destination_filename), exist_ok=True)\n",
    "        try:\n",
    "            with open(destination_filename, 'wb') as local_file:\n",
    "                ftp.retrbinary(f'RETR {ftp_path}', local_file.write)\n",
    "        except Exception as e:\n",
    "            print(f\"Error downloading {ftp_url}: {e}\")\n",
    "\n",
    "def get_species_name(tax_id):\n",
    "    \"\"\" Fetch species name using taxonomic ID from Ensembl \"\"\"\n",
    "    url = f\"https://rest.ensembl.org/taxonomy/id/{tax_id}?content-type=application/json\"\n",
    "    response = requests.get(url)\n",
    "    return response.json()['scientific_name']\n",
    "\n",
    "def find_matching_files(ftp, path, patterns):\n",
    "    \"\"\" Find files matching given patterns \"\"\"\n",
    "    ftp.cwd(path)\n",
    "    files = ftp.nlst()\n",
    "    matched_files = [file for file in files if any(pattern in file for pattern in patterns)]\n",
    "    return matched_files\n",
    "\n",
    "def download_ensembl_files(tax_id, path):\n",
    "    species_name = get_species_name(tax_id)\n",
    "    species_path = species_name.lower().replace(' ', '_')\n",
    "\n",
    "    # Fetch genome information from Ensembl\n",
    "    url = f\"https://rest.ensembl.org/info/genomes/{species_path}?content-type=application/json\"\n",
    "    response = requests.get(url)\n",
    "    data = response.json()\n",
    "    assembly_name = data['assembly_name']\n",
    "\n",
    "    # Prepare directory name\n",
    "    dir_name = sanitize_filename(f\"{tax_id}-{species_name}\")\n",
    "    dir_path = os.path.join(path, dir_name)\n",
    "    os.makedirs(dir_path, exist_ok=True)\n",
    "\n",
    "    # Prepare FTP paths\n",
    "    ensembl_ftp_base = \"ftp.ensembl.org\"\n",
    "    fasta_ftp_dir = f\"/pub/release-110/fasta/{species_path}/dna/\"\n",
    "    gtf_ftp_dir = f\"/pub/release-110/gtf/{species_path}/\"\n",
    "\n",
    "    with FTP(ensembl_ftp_base) as ftp:\n",
    "        ftp.login()\n",
    "\n",
    "        # Find matching FASTA and GTF files\n",
    "        fa_patterns = [\".dna_sm.toplevel.fa.gz\"]\n",
    "        gtf_patterns = [\".gtf.gz\"]\n",
    "        unwanted_patterns = [\"chr\", \"hapl\", \"abinitio\"]\n",
    "\n",
    "        fasta_files = find_matching_files(ftp, fasta_ftp_dir, fa_patterns)\n",
    "        gtf_files = find_matching_files(ftp, gtf_ftp_dir, gtf_patterns)\n",
    "        fasta_files = [f for f in fasta_files if not any(p in f for p in unwanted_patterns)]\n",
    "        gtf_files = [f for f in gtf_files if not any(p in f for p in unwanted_patterns)]\n",
    "\n",
    "        # Download files\n",
    "        if fasta_files:\n",
    "            download_ftp_file(f\"ftp://{ensembl_ftp_base}{fasta_ftp_dir}{fasta_files[0]}\", os.path.join(dir_path, fasta_files[0]))\n",
    "        if gtf_files:\n",
    "            download_ftp_file(f\"ftp://{ensembl_ftp_base}{gtf_ftp_dir}{gtf_files[0]}\", os.path.join(dir_path, gtf_files[0]))\n",
    "\n",
    "\n",
    "species_list = species_dict.keys()\n",
    "#species_list=[str(x) for x in species_list]\n",
    "for species in species_list:\n",
    "    download_ensembl_files(species, '/home/matthew.schmitz/Matthew/genome/ensembl')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab421dca-cc5e-4c46-9ae0-1e5f5fe114b4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genomes",
   "language": "python",
   "name": "genomes"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
